// @ucloud-fe/react-components/lib/components/Breadcrumb/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.BreadcrumbWrap = exports.SeparatorWrap = exports.ItemWrapA = exports.ItemWrapSpan = exports.BackButtonWrap = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _style = require('../../../style');

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

var _Button = require('../../../components/Button');

var _Button2 = _interopRequireDefault(_Button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var textStyleMixin = (0, _styledComponents.css)(['font-size:14px;margin-right:6px;', ';'], _style.inlineBlockWithVerticalMixin);

var BackButtonWrap = exports.BackButtonWrap = (0, _styledComponents2.default)(_Button2.default).attrs({
    icon: 'left',
    size: 'sm',
    styleType: 'border-gray'
}).withConfig({
    displayName: 'style__BackButtonWrap',
    componentId: 'sc-8t6fv0-0'
})(['margin-right:16px;padding:0 5px;']);

var itemWrapMixin = function itemWrapMixin(_ref) {
    var _ref$theme = _ref.theme,
        colorMap = _ref$theme.colorMap,
        colorList = _ref$theme.colorList,
        disabled = _ref.disabled,
        current = _ref.current,
        noAction = _ref.noAction;
    return (0, _styledComponents.css)(['cursor:pointer;text-decoration:none;', ';', ';', ';', ';'], textStyleMixin, noAction && (0, _styledComponents.css)(['pointer-events:none;color:', ' !important;cursor:default;'], colorMap.default.text), current && (0, _styledComponents.css)(['pointer-events:none;color:', ' !important;font-weight:bold;'], colorList.black), disabled && (0, _styledComponents.css)(['pointer-events:none;color:', ' !important;'], colorMap.disabled.text));
};

var ItemWrapSpan = exports.ItemWrapSpan = _styledComponents2.default.span.withConfig({
    displayName: 'style__ItemWrapSpan',
    componentId: 'sc-8t6fv0-1'
})(['', ';'], itemWrapMixin);

var ItemWrapA = exports.ItemWrapA = _styledComponents2.default.a.withConfig({
    displayName: 'style__ItemWrapA',
    componentId: 'sc-8t6fv0-2'
})(function (_ref2) {
    var colorMap = _ref2.theme.colorMap;
    return (0, _styledComponents.css)(['', ';&,&:hover,&:visited,&:link,&:active{color:', ';}'], itemWrapMixin, colorMap.default.text);
});

var SeparatorWrap = exports.SeparatorWrap = _styledComponents2.default.span.withConfig({
    displayName: 'style__SeparatorWrap',
    componentId: 'sc-8t6fv0-3'
})(['', ';font-weight:bold;cursor:default;'], textStyleMixin);
/* stylelint-disable no-descending-specificity */
var styleTypeMixin = function styleTypeMixin(_ref3) {
    var styleType = _ref3.styleType,
        colorMap = _ref3.theme.colorMap;
    return {
        'block-hover': (0, _styledComponents.css)([':hover{', '{color:', ';}', '{&,&:hover,&:visited,&:link,&:active{color:', ';}}}'], ItemWrapSpan, colorMap.active.text, ItemWrapA, colorMap.active.text),
        active: (0, _styledComponents.css)(['', '{color:', ';}', '{&,&:hover,&:visited,&:link,&:active{color:', ';}}'], ItemWrapSpan, colorMap.active.text, ItemWrapA, colorMap.active.text),
        hover: (0, _styledComponents.css)(['', ',', '{:hover{color:', ';}}'], ItemWrapSpan, ItemWrapA, colorMap.active.text)
    }[styleType];
};
/* stylelint-enable no-descending-specificity */

var BreadcrumbWrap = exports.BreadcrumbWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__BreadcrumbWrap',
    componentId: 'sc-8t6fv0-4'
})(function (_ref4) {
    var _ref4$theme = _ref4.theme,
        colorMap = _ref4$theme.colorMap,
        fontSize = _ref4$theme.fontSize;
    return (0, _styledComponents.css)(['font-size:', ';color:', ';vertical-align:baseline;', ';'], fontSize, colorMap.default.text, styleTypeMixin);
});

(0, _addDefaultThemeProps2.default)(ItemWrapA, ItemWrapSpan, BreadcrumbWrap);
