// @ucloud-fe/react-components/lib/components/Transfer/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TransferWrap = exports.disabledCls = exports.searchClearBtnCls = exports.listWrapCls = exports.tipWrapCls = exports.searchCls = exports.actionWrapCls = exports.actionCls = exports.titleCls = exports.footerCls = exports.partContentCls = exports.partWrapCls = exports.prefixCls = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _prefixCls = _config2.default.prefixCls;
var prefixCls = exports.prefixCls = _prefixCls + '-transfer';
var partWrapCls = exports.partWrapCls = prefixCls + '-part-wrap';
var partContentCls = exports.partContentCls = prefixCls + '-part-content';
var footerCls = exports.footerCls = prefixCls + '-footer';
var titleCls = exports.titleCls = prefixCls + '-title';
var actionCls = exports.actionCls = prefixCls + '-action';
var actionWrapCls = exports.actionWrapCls = prefixCls + '-action-wrap';
var searchCls = exports.searchCls = prefixCls + '-search-wrap';
var tipWrapCls = exports.tipWrapCls = prefixCls + '-tip-wrap';
var listWrapCls = exports.listWrapCls = prefixCls + '-list-wrap';
var searchClearBtnCls = exports.searchClearBtnCls = prefixCls + '-clear-search-btn';
var disabledCls = exports.disabledCls = prefixCls + '-disabled';

var TransferWrap = exports.TransferWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__TransferWrap',
    componentId: 'sc-1to7aw9-0'
})(function (_ref) {
    var DT = _ref.theme.designTokens;
    return (0, _styledComponents.css)(['display:flex;align-items:stretch;justify-content:flex-start;.', '{display:flex;flex:1 1 50%;flex-direction:column;}.', '{margin:12px;}.', '{color:', ';font-size:12px;line-height:20px;padding:6px 4px;}.', '{border-radius:2px;width:100%;border:1px solid ', ';flex:auto;display:flex;flex-direction:column;}.', '{.', ',.', '{border-color:', ';}.', '{color:', ';}}.', '{padding:12px;display:flex;align-items:center;flex-direction:column;align-self:center;.', '+.', '{margin-top:8px;}}.', '{margin:50px;padding:8px;flex:auto;}.', '{flex:auto;}.', '{border-top:1px solid ', ';padding:8px;}.', '{margin-left:4px;cursor:pointer;color:', ';}'], partWrapCls, searchCls, titleCls, DT.T_COLOR_TEXT_DEFAULT_DARK, partContentCls, DT.T_COLOR_LINE_DEFAULT_DARK, disabledCls, partContentCls, footerCls, DT.T_COLOR_LINE_DISABLED_LIGHT, titleCls, DT.T_COLOR_TEXT_DISABLED, actionWrapCls, actionCls, actionCls, tipWrapCls, listWrapCls, footerCls, DT.T_COLOR_LINE_DEFAULT_DARK, searchClearBtnCls, DT.T_COLOR_TEXT_PRIMARY_DEFAULT);
});

(0, _addDefaultThemeProps2.default)(TransferWrap);
