// @ucloud-fe/react-components/lib/components/DatePicker/DatePicker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.Size = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _NumberInput = require('../../components/NumberInput');

var _NumberInput2 = _interopRequireDefault(_NumberInput);

var _Calendar = require('../../components/Calendar/Calendar');

var _Calendar2 = _interopRequireDefault(_Calendar);

var _placements = require('../../components/Popover/placements');

var _placements2 = _interopRequireDefault(_placements);

var _uncontrolled = require('../../decorators/uncontrolled');

var _uncontrolled2 = _interopRequireDefault(_uncontrolled);

var _globalAnimation = require('../../style/globalAnimation');

var _ContainerContext = require('../../components/Popover/ContainerContext');

var _utils = require('./utils');

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Size = exports.Size = ['sm', 'md', 'lg'];

var DatePicker = (_dec = (0, _uncontrolled2.default)({}), _dec(_class = (_temp2 = _class2 = function (_Component) {
    _inherits(DatePicker, _Component);

    function DatePicker() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, DatePicker);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = DatePicker.__proto__ || Object.getPrototypeOf(DatePicker)).call.apply(_ref, [this].concat(args))), _this), _this.handleChange = function (value) {
            var _this$props = _this.props,
                onChange = _this$props.onChange,
                rules = _this$props.rules;


            if ((0, _utils.isDateDisabled)(value, _this.props.value, rules)) {
                return;
            }
            value = (0, _utils.getValidDate)(value, rules);

            onChange(value);
        }, _this.renderTimePicker = function (value, display) {
            var _this$props2 = _this.props,
                size = _this$props2.size,
                disabled = _this$props2.disabled;
            var hour = display.hour,
                minute = display.minute,
                second = display.second;

            var picker = [];
            if (hour !== false) {
                picker.push(_react2.default.createElement(_NumberInput2.default, {
                    size: size,
                    disabled: disabled,
                    key: 'hour-input',
                    value: value.hour(),
                    styleType: 'pagination',
                    hideHandler: true,
                    onNumberChange: function onNumberChange(hour) {
                        return _this.handleChange((0, _moment2.default)(value).set('hour', hour));
                    }
                }));
            }
            if (minute !== false) {
                if (picker.length) {
                    picker.push(_react2.default.createElement(
                        _style.TimeSeparator,
                        { key: 'separator-1' },
                        ':'
                    ));
                }
                picker.push(_react2.default.createElement(_NumberInput2.default, {
                    size: size,
                    disabled: disabled,
                    key: 'minute-input',
                    value: value.minute(),
                    styleType: 'pagination',
                    hideHandler: true,
                    onNumberChange: function onNumberChange(minute) {
                        return _this.handleChange((0, _moment2.default)(value).set('minute', minute));
                    }
                }));
            }
            if (second !== false) {
                if (picker.length) {
                    picker.push(_react2.default.createElement(
                        _style.TimeSeparator,
                        { key: 'separator-2' },
                        ':'
                    ));
                }
                picker.push(_react2.default.createElement(_NumberInput2.default, {
                    size: size,
                    disabled: disabled,
                    key: 'second-input',
                    value: value.second(),
                    styleType: 'pagination',
                    hideHandler: true,
                    onNumberChange: function onNumberChange(second) {
                        return _this.handleChange((0, _moment2.default)(value).set('second', second));
                    }
                }));
            }
            return _react2.default.createElement(
                _style.TimeWrap,
                null,
                picker
            );
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(DatePicker, [{
        key: 'render',
        value: function render() {
            var _this2 = this;

            var _props = this.props,
                rules = _props.rules,
                _v = _props.value,
                defaultValue = _props.defaultValue,
                _props$display = _props.display,
                display = _props$display === undefined ? {} : _props$display,
                size = _props.size,
                onChange = _props.onChange,
                zIndex = _props.zIndex,
                getCalendarContainer = _props.getCalendarContainer,
                rest = _objectWithoutProperties(_props, ['rules', 'value', 'defaultValue', 'display', 'size', 'onChange', 'zIndex', 'getCalendarContainer']);

            var _display$date = display.date,
                date = _display$date === undefined ? {} : _display$date;

            var value = (0, _moment2.default)(_v);

            return _react2.default.createElement(
                _ContainerContext.Consumer,
                null,
                function () {
                    var _ref2 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
                        getPopupContainer = _ref2.getPopupContainer;

                    return _react2.default.createElement(
                        _style.PickerContainer,
                        rest,
                        date.show !== false && _react2.default.createElement(
                            _style.PickerWrap,
                            {
                                prefixCls: _style.pickerPrefixCls,
                                transitionName: _globalAnimation.animationPrefixCls + '-fade',
                                calendar: _react2.default.createElement(_Calendar2.default, { rules: rules }),
                                getCalendarContainer: getCalendarContainer || getPopupContainer || function (triggerNode) {
                                    return triggerNode.parentNode;
                                },
                                value: value,
                                align: _placements2.default.bottomLeft,
                                onChange: onChange,
                                zIndex: zIndex
                            },
                            function (_ref3) {
                                var value = _ref3.value;

                                return _react2.default.createElement(
                                    _style.DateWrap,
                                    { size: size },
                                    _react2.default.createElement(
                                        _style.DateSpan,
                                        null,
                                        value.format(date.format || 'YYYY-MM-DD')
                                    ),
                                    _react2.default.createElement(_style.PickerIcon, { type: 'calendar', color: 'blue' })
                                );
                            }
                        ),
                        _this2.renderTimePicker(value, display)
                    );
                }
            );
        }
    }]);

    return DatePicker;
}(_react.Component), _class2.propTypes = {
    /** 当前值，受控 */
    value: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.number]),
    /** 默认值，非受控 */
    defaultValue: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.number]),
    /** 修改回调，返回moment对象 */
    onChange: _propTypes2.default.func,
    /** 自定义规则，参考Calendar */
    rules: _propTypes2.default.object,
    /** 控件尺寸 */
    size: _propTypes2.default.oneOf(Size),
    /** 展示格式 */
    display: _propTypes2.default.shape({
        date: _propTypes2.default.shape({
            format: _propTypes2.default.string,
            display: _propTypes2.default.bool
        }),
        hour: _propTypes2.default.bool,
        minute: _propTypes2.default.bool,
        second: _propTypes2.default.bool
    }),
    /** 是否禁用 */
    disabled: _propTypes2.default.bool,
    /** 弹出层的z-index */
    zIndex: _propTypes2.default.number,
    /** 弹出层容器，参考popover.getPopupContainer */
    getCalendarContainer: _propTypes2.default.func
}, _class2.defaultProps = {
    onChange: function onChange() {},
    size: 'md',
    zIndex: 100
}, _temp2)) || _class);
exports.default = DatePicker;


DatePicker.Size = Size;
