// @ucloud-fe/react-components/lib/components/DatePicker/Month.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _Month = require('../../components/Calendar/Month');

var _Month2 = _interopRequireDefault(_Month);

var _placements = require('../../components/Popover/placements');

var _placements2 = _interopRequireDefault(_placements);

var _uncontrolled = require('../../decorators/uncontrolled');

var _uncontrolled2 = _interopRequireDefault(_uncontrolled);

var _globalAnimation = require('../../style/globalAnimation');

var _ContainerContext = require('../../components/Popover/ContainerContext');

var _style = require('./style');

var _DatePicker = require('./DatePicker');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Month = (_dec = (0, _uncontrolled2.default)({}), _dec(_class = (_temp = _class2 = function (_Component) {
    _inherits(Month, _Component);

    function Month() {
        _classCallCheck(this, Month);

        return _possibleConstructorReturn(this, (Month.__proto__ || Object.getPrototypeOf(Month)).apply(this, arguments));
    }

    _createClass(Month, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                rules = _props.rules,
                _v = _props.value,
                defaultValue = _props.defaultValue,
                size = _props.size,
                _props$display = _props.display,
                display = _props$display === undefined ? {} : _props$display,
                onChange = _props.onChange,
                zIndex = _props.zIndex,
                getCalendarContainer = _props.getCalendarContainer,
                rest = _objectWithoutProperties(_props, ['rules', 'value', 'defaultValue', 'size', 'display', 'onChange', 'zIndex', 'getCalendarContainer']);

            var _display$date = display.date,
                date = _display$date === undefined ? {} : _display$date;

            var value = (0, _moment2.default)(_v);

            return _react2.default.createElement(
                _ContainerContext.Consumer,
                null,
                function () {
                    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
                        getPopupContainer = _ref.getPopupContainer;

                    return _react2.default.createElement(
                        _style.PickerContainer,
                        _extends({ isMonth: true }, rest),
                        _react2.default.createElement(
                            _style.PickerWrap,
                            {
                                prefixCls: _style.monthPickerPrefixCls,
                                transitionName: _globalAnimation.animationPrefixCls + '-fade',
                                calendar: _react2.default.createElement(_Month2.default, { rules: rules }),
                                getCalendarContainer: getCalendarContainer || getPopupContainer || function (triggerNode) {
                                    return triggerNode.parentNode;
                                },
                                value: value,
                                align: _placements2.default.bottomLeft,
                                onChange: onChange,
                                zIndex: zIndex,
                                isMonth: true
                            },
                            function (_ref2) {
                                var value = _ref2.value;

                                return _react2.default.createElement(
                                    _style.DateWrap,
                                    { size: size },
                                    _react2.default.createElement(
                                        _style.DateSpan,
                                        null,
                                        value.format(date.format || 'YYYY-MM')
                                    ),
                                    _react2.default.createElement(_style.PickerIcon, { type: 'calendar', color: 'blue' })
                                );
                            }
                        )
                    );
                }
            );
        }
    }]);

    return Month;
}(_react.Component), _class2.propTypes = {
    /** 当前值，受控 */
    value: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.number]),
    /** 默认值，非受控 */
    defaultValue: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.number]),
    /** 修改回调，返回moment对象 */
    onChange: _propTypes2.default.func,
    /** 自定义规则，参考Calendar */
    rules: _propTypes2.default.object,
    /** 展示格式 */
    display: _propTypes2.default.shape({
        date: _propTypes2.default.shape({
            format: _propTypes2.default.string
        })
    }),
    /** 控件尺寸 */
    size: _propTypes2.default.oneOf(_DatePicker.Size),
    /** 是否禁用 */
    disabled: _propTypes2.default.bool,
    /** 弹出层的z-index */
    zIndex: _propTypes2.default.number,
    /** 弹出层容器，参考popover.getPopupContainer */
    getCalendarContainer: _propTypes2.default.func
}, _class2.defaultProps = {
    onChange: function onChange() {},
    size: 'md',
    zIndex: 100
}, _temp)) || _class);
exports.default = Month;
