// @ucloud-fe/react-components/lib/components/EditableTable/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.AddBar = exports.RemoveBtn = exports.AddTip = exports.AddBtn = exports.prefixCls = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

var _SvgIcon = require('../../../components/SvgIcon');

var _SvgIcon2 = _interopRequireDefault(_SvgIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _prefixCls = _config2.default.prefixCls;
var prefixCls = exports.prefixCls = _prefixCls + '-editable-table';

var AddBtn = exports.AddBtn = (0, _styledComponents2.default)(_SvgIcon2.default)(function () {
    return (0, _styledComponents.css)(['']);
});

var AddTip = exports.AddTip = _styledComponents2.default.span.withConfig({
    displayName: 'style__AddTip',
    componentId: 'uiydpr-0'
})(function (_ref) {
    var DT = _ref.theme.designTokens;
    return (0, _styledComponents.css)(['font-size:12px;color:', ';line-height:20px;margin-left:8px;vertical-align:middle;'], DT.T_COLOR_TEXT_DEFAULT_DARK);
});

var RemoveBtn = exports.RemoveBtn = (0, _styledComponents2.default)(_SvgIcon2.default)(function (_ref2) {
    var DT = _ref2.theme.designTokens;
    return (0, _styledComponents.css)(['cursor:pointer;fill:', ';:hover{fill:', ';}'], DT.T_COLOR_TEXT_DEFAULT_LIGHT, DT.T_COLOR_TEXT_PRIMARY_DEFAULT);
});

var AddBar = exports.AddBar = _styledComponents2.default.div.withConfig({
    displayName: 'style__AddBar',
    componentId: 'uiydpr-1'
})(function (_ref3) {
    var DT = _ref3.theme.designTokens,
        disabled = _ref3.disabled;
    return (0, _styledComponents.css)(['width:100%;height:60px;background:', ';color:', ';cursor:pointer;text-align:center;line-height:60px;', '{fill:', ';}:hover{background:', ';', '{fill:', ';}}', ''], DT.T_COLOR_BG_DEFAULT_LIGHT, DT.T_COLOR_TEXT_DEFAULT_DARK, AddBtn, DT.T_COLOR_TEXT_DEFAULT_DARK, DT.T_COLOR_BG_DEFAULT_HOVER, AddBtn, DT.T_COLOR_TEXT_PRIMARY_DEFAULT, disabled && (0, _styledComponents.css)(['&&&{background:', ';color:', ';cursor:not-allowed;', '{color:', ';}', '{fill:', ';}}'], DT.T_COLOR_BG_DISABLED_LIGHT, DT.T_COLOR_TEXT_DISABLED, AddTip, DT.T_COLOR_TEXT_DISABLED, AddBtn, DT.T_COLOR_TEXT_DISABLED));
});

(0, _addDefaultThemeProps2.default)(AddBar, AddTip, RemoveBtn);
