// @ucloud-fe/react-components/lib/components/Message/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.MessageContentWrap = exports.MessageWrap = exports.FooterWrap = exports.ContentWrap = exports.TitleWrap = exports.CloseIconWrap = exports.IconWrap = exports.TipIcon = exports.animationName = exports.animationDuration = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _reactTransitionGroup = require('react-transition-group');

var _SvgIcon = require('../../../components/SvgIcon');

var _SvgIcon2 = _interopRequireDefault(_SvgIcon);

var _animation = require('../../../style/animation');

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var animationDuration = exports.animationDuration = 500;
var animationName = exports.animationName = 'uc-fe-animation-fade';

var colorMap = {
    success: 'T_COLOR_TEXT_SUCCESS',
    warning: 'T_COLOR_TEXT_WARNING',
    error: 'T_COLOR_TEXT_ERROR',
    loading: 'T_COLOR_TEXT_PRIMARY_DEFAULT',
    default: 'T_COLOR_TEXT_PRIMARY_DEFAULT'
};

var TipIcon = exports.TipIcon = (0, _styledComponents2.default)(_SvgIcon2.default)(function (_ref) {
    var styleType = _ref.styleType,
        DT = _ref.theme.designTokens;
    return (0, _styledComponents.css)(['fill:', ';'], DT[colorMap[styleType]]);
});

var IconWrap = exports.IconWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__IconWrap',
    componentId: 'sc-1hnl1v6-0'
})(['font-size:15px;width:16px;height:16px;text-align:center;line-height:16px;position:absolute;top:16px;left:16px;']);
var CloseIconWrap = exports.CloseIconWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__CloseIconWrap',
    componentId: 'sc-1hnl1v6-1'
})(['font-size:15px;width:16px;height:16px;text-align:center;line-height:16px;position:absolute;top:16px;right:12px;cursor:pointer;']);

var TitleWrap = exports.TitleWrap = _styledComponents2.default.h3.withConfig({
    displayName: 'style__TitleWrap',
    componentId: 'sc-1hnl1v6-2'
})(function (_ref2) {
    var DT = _ref2.theme.designTokens;

    return (0, _styledComponents.css)(['padding-right:36px;padding-left:44px;word-break:break-all;font-size:14px;color:', ';line-height:22px;'], DT.T_COLOR_TEXT_DEFAULT_DARK);
});

var ContentWrap = exports.ContentWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__ContentWrap',
    componentId: 'sc-1hnl1v6-3'
})(function (_ref3) {
    var DT = _ref3.theme.designTokens;

    return (0, _styledComponents.css)(['padding-right:36px;padding-left:44px;word-break:break-all;font-size:12px;color:', ';line-height:20px;'], DT.T_COLOR_TEXT_DEFAULT_LIGHT);
});

var FooterWrap = exports.FooterWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__FooterWrap',
    componentId: 'sc-1hnl1v6-4'
})(['text-align:right;margin-right:16px;margin-top:16px;']);

var MessageWrap = exports.MessageWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__MessageWrap',
    componentId: 'sc-1hnl1v6-5'
})(function (_ref4) {
    var DT = _ref4.theme.designTokens;

    return (0, _styledComponents.css)(['box-sizing:border-box;width:340px;min-height:52px;padding:16px 0;border-radius:4px;box-shadow:', ';background:', ';position:relative;', '+', '{margin-top:8px;}'], DT.T_SHADOW_BLOCK_DEFAULT_MD, DT.T_CARD_COLOR_BG_DEFAULT, TitleWrap, ContentWrap);
});

var MessageContentWrap = exports.MessageContentWrap = (0, _styledComponents2.default)(_reactTransitionGroup.TransitionGroup).withConfig({
    displayName: 'style__MessageContentWrap',
    componentId: 'sc-1hnl1v6-6'
})(['position:fixed;top:0;width:100%;height:0px;overflow:visible;z-index:1060;.', '-enter,.', '-appear,.', '-leave,.', '-exit{animation-duration:', 'ms;animation-fill-mode:both;}.', '-enter,.', '-appear{animation-name:', ';}.', '-leave,.', '-exit{animation-name:', ';}', '{position:relative;margin-bottom:16px;margin-right:16px;margin-left:auto;}'], animationName, animationName, animationName, animationName, animationDuration, animationName, animationName, _animation.fadeIn, animationName, animationName, _animation.fadeOut, /*sc-sel*/MessageWrap);

(0, _addDefaultThemeProps2.default)(MessageWrap, TipIcon, TitleWrap, ContentWrap);
