// @ucloud-fe/react-components/lib/components/Menu/Item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Checkbox = require('../../components/Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Item = (_temp = _class = function (_PureComponent) {
    _inherits(Item, _PureComponent);

    function Item() {
        _classCallCheck(this, Item);

        return _possibleConstructorReturn(this, (Item.__proto__ || Object.getPrototypeOf(Item)).apply(this, arguments));
    }

    _createClass(Item, [{
        key: 'render',
        value: function render() {
            /* eslint-disable no-unused-vars */
            var _props = this.props,
                itemKey = _props.itemKey,
                disabled = _props.disabled,
                uid = _props.uid,
                selected = _props.selected,
                multiple = _props.multiple,
                onSelect = _props.onSelect,
                children = _props.children,
                className = _props.className,
                rest = _objectWithoutProperties(_props, ['itemKey', 'disabled', 'uid', 'selected', 'multiple', 'onSelect', 'children', 'className']);
            /* eslint-enable no-unused-vars */

            return _react2.default.createElement(
                'div',
                _extends({
                    className: (0, _classnames2.default)(_style.itemCls, selected && _style.selectedCls, disabled && _style.disabledCls, className),
                    selected: selected,
                    disabled: disabled,
                    onClick: function onClick() {
                        return !disabled && onSelect(multiple ? !selected : true, uid);
                    }
                }, rest),
                multiple ? _react2.default.createElement(
                    _Checkbox2.default,
                    { className: _style.checkboxCls, checked: selected, disabled: disabled, size: 'lg' },
                    children
                ) : children
            );
        }
    }]);

    return Item;
}(_react.PureComponent), _class.propTypes = {
    /** item的唯一键 */
    itemKey: _propTypes2.default.any,
    /** 是否禁用 */
    disabled: _propTypes2.default.bool,
    /** @ignore */
    uid: _propTypes2.default.string,
    /** @ignore */
    selected: _propTypes2.default.bool,
    /** @ignore */
    multiple: _propTypes2.default.bool,
    /** @ignore */
    onSelect: _propTypes2.default.func,
    /** @ignore */
    children: _propTypes2.default.node,
    /** @ignore */
    className: _propTypes2.default.string
}, _class.getItemKey = function (item) {
    return item.props.itemKey;
}, _temp);


Item.isMenuItem = true;

exports.default = Item;
