// @ucloud-fe/react-components/lib/components/Form/Item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Item = function Item(_ref) {
    var label = _ref.label,
        children = _ref.children,
        labelCol = _ref.labelCol,
        controllerCol = _ref.controllerCol,
        rest = _objectWithoutProperties(_ref, ['label', 'children', 'labelCol', 'controllerCol']);

    return _react2.default.createElement(
        _style.ItemWrap,
        rest,
        _react2.default.createElement(
            _style.LabelWrap,
            labelCol,
            label
        ),
        _react2.default.createElement(
            _style.ControllerWrap,
            controllerCol,
            children
        )
    );
};

var colShape = {
    /** 栅格占位格数 */
    span: _propTypes2.default.number,
    /** 栅格向右偏移格数，偏移占位 */
    offset: _propTypes2.default.number,
    /** 栅格向左偏移格数，偏移不占位 */
    pull: _propTypes2.default.number,
    /** 栅格向右偏移格数，偏移不占位 */
    push: _propTypes2.default.number
};

Item.propTypes = {
    /** 表单项标签 */
    label: _propTypes2.default.node,
    /** @ignore */
    children: _propTypes2.default.node,
    /** 标签的col配置 */
    labelCol: _propTypes2.default.shape(colShape),
    /** 控件的col配置 */
    controllerCol: _propTypes2.default.shape(colShape),
    /** @ignore */
    className: _propTypes2.default.string
};

exports.default = Item;
