// @ucloud-fe/react-components/lib/components/EditableList/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ListWrap = exports.actionCls = exports.itemCls = exports.liCls = exports.liGridCls = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _prefixCls = _config2.default.prefixCls;

var prefixCls = _prefixCls + '-editable-list';

var liGridCls = exports.liGridCls = prefixCls + '-li-grid';
var liCls = exports.liCls = prefixCls + '-li';
var itemCls = exports.itemCls = prefixCls + '-item';
var actionCls = exports.actionCls = prefixCls + '-action';

var ListWrap = exports.ListWrap = (0, _styledComponents2.default)('div').withConfig({
    displayName: 'style__ListWrap',
    componentId: 'sc-9ofvqc-0'
})(['> .', '+.', '{margin-top:16px;}> .', '+.', '{margin-top:16px;}.', '{.', ',.', '{display:inline-block;}.', '{margin-left:24px;}}'], liCls, liCls, liGridCls, liGridCls, liCls, itemCls, actionCls, actionCls);
