// @ucloud-fe/react-components/lib/components/Notice/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.NoticeWrap = exports.CloseIcon = exports.CloseWrap = exports.ActionWrap = exports.ContentWrap = exports.NoticeIcon = exports.NoticeIconWrap = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _Icon = require('../../../components/Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var map = {
    default: 'info',
    success: 'success',
    warning: 'warning',
    error: 'error',
    info: 'success',
    disabled: 'disabled'
};

var NoticeIconWrap = exports.NoticeIconWrap = _styledComponents2.default.span.withConfig({
    displayName: 'style__NoticeIconWrap',
    componentId: 'hces35-0'
})(['display:table-cell;font-size:15px;width:15px;color:inherit;padding-right:8px;padding-top:2px;']);

var NoticeIcon = exports.NoticeIcon = (0, _styledComponents2.default)(_Icon2.default).withConfig({
    displayName: 'style__NoticeIcon',
    componentId: 'hces35-1'
})(['']);

var ContentWrap = exports.ContentWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__ContentWrap',
    componentId: 'hces35-2'
})(['display:table-cell;vertical-align:middle;']);

var ActionWrap = exports.ActionWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__ActionWrap',
    componentId: 'hces35-3'
})(['display:table-cell;padding-left:8px;white-space:nowrap;text-align:right;']);

var CloseWrap = exports.CloseWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__CloseWrap',
    componentId: 'hces35-4'
})(['display:table-cell;padding-left:8px;white-space:nowrap;text-align:right;width:12px;']);

var CloseIcon = exports.CloseIcon = (0, _styledComponents2.default)(_Icon2.default).withConfig({
    displayName: 'style__CloseIcon',
    componentId: 'hces35-5'
})(['cursor:pointer;']);

var themeMixin = function themeMixin(_ref) {
    var styleType = _ref.styleType,
        DT = _ref.theme.designTokens;

    var style = map[styleType];
    var colorMap = {
        info: {
            color: DT.T_COLOR_TEXT_DEFAULT_DARK,
            border: DT.T_COLOR_LINE_NOTICE_LIGHT,
            background: DT.T_COLOR_BG_NOTICE_LIGHT,
            icon: DT.T_COLOR_TEXT_PRIMARY_DEFAULT
        },
        success: {
            color: DT.T_COLOR_TEXT_DEFAULT_DARK,
            border: DT.T_COLOR_LINE_SUCCESS_LIGHT,
            background: DT.T_COLOR_BG_SUCCESS_LIGHT,
            icon: DT.T_COLOR_TEXT_SUCCESS
        },
        warning: {
            color: DT.T_COLOR_TEXT_DEFAULT_DARK,
            border: DT.T_COLOR_LINE_WARNING_LIGHT,
            background: DT.T_COLOR_BG_WARNING_LIGHT,
            icon: DT.T_COLOR_TEXT_WARNING
        },
        error: {
            color: DT.T_COLOR_TEXT_DEFAULT_DARK,
            border: DT.T_COLOR_LINE_ERROR_LIGHT,
            background: DT.T_COLOR_BG_ERROR_LIGHT,
            icon: DT.T_COLOR_TEXT_ERROR
        },
        disabled: {
            color: DT.T_COLOR_TEXT_DISABLED,
            border: DT.T_COLOR_LINE_DISABLED_LIGHT,
            background: DT.T_COLOR_BG_DISABLED,
            icon: DT.T_COLOR_TEXT_DISABLED
        }
    }[style];
    return (0, _styledComponents.css)(['color:', ';border:', ' solid ', ';background:', ';', '{color:', ';}', '{color:', ';}'], colorMap.color, DT.T_LINE_WIDTH_BASE, colorMap.border, colorMap.background, NoticeIcon, colorMap.icon, ActionWrap, DT.T_COLOR_TEXT_DEFAULT_LIGHT);
};

var NoticeWrap = exports.NoticeWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__NoticeWrap',
    componentId: 'hces35-6'
})(['display:table;position:relative;box-sizing:border-box;width:100%;padding:10px 16px;margin:0;border-radius:1px;line-height:18px;overflow:hidden;', ';'], themeMixin);
(0, _addDefaultThemeProps2.default)(NoticeWrap);
