// @ucloud-fe/react-components/lib/components/Input/Search.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _KeyCode = require('../../interfaces/KeyCode');

var _KeyCode2 = _interopRequireDefault(_KeyCode);

var _Input = require('./Input');

var _Input2 = _interopRequireDefault(_Input);

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Search = (_temp2 = _class = function (_Component) {
    _inherits(Search, _Component);

    function Search() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Search);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Search.__proto__ || Object.getPrototypeOf(Search)).call.apply(_ref, [this].concat(args))), _this), _this.onSearch = function () {
            var _this$props = _this.props,
                onSearch = _this$props.onSearch,
                disabled = _this$props.disabled;

            !disabled && onSearch(_this.input.input.value);
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Search, [{
        key: 'render',
        value: function render() {
            var _this2 = this;

            // eslint-disable-next-line no-unused-vars
            var _props = this.props,
                onSearch = _props.onSearch,
                _onKeyDown = _props.onKeyDown,
                rest = _objectWithoutProperties(_props, ['onSearch', 'onKeyDown']);

            return _react2.default.createElement(_Input2.default, _extends({}, rest, {
                onKeyDown: function onKeyDown(e) {
                    if (e.keyCode === _KeyCode2.default['ENTER']) {
                        _this2.onSearch();
                        e.preventDefault();
                    }
                    if (_onKeyDown) {
                        _onKeyDown(e);
                    }
                },
                ref: function ref(_ref2) {
                    return _this2.input = _ref2;
                },
                suffix: _react2.default.createElement(_style.SearchIcon, { type: 'search', onClick: this.onSearch })
            }));
        }
    }]);

    return Search;
}(_react.Component), _class.propTypes = {
    /** 搜索回调 */
    onSearch: _propTypes2.default.func,
    /** @ignore */
    onKeyDown: _propTypes2.default.func,
    /** @ignore */
    disabled: _propTypes2.default.bool
}, _class.defaultProps = {
    onSearch: function onSearch() {}
}, _temp2);
exports.default = Search;
