// @ucloud-fe/react-components/lib/components/LocaleProvider/localeConsumerDecorator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _runtime = require('./locale/runtime');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var localeConsumerDecorator = function localeConsumerDecorator(_ref) {
    var _ref$defaultLocale = _ref.defaultLocale,
        defaultLocale = _ref$defaultLocale === undefined ? {} : _ref$defaultLocale,
        localeName = _ref.localeName,
        _ref$publicFn = _ref.publicFn,
        publicFn = _ref$publicFn === undefined ? [] : _ref$publicFn,
        requireRuntimeLocale = _ref.requireRuntimeLocale;
    return function (Child) {
        var _class, _temp;

        var LocalConsumerWrappedComponent = (_temp = _class = function (_Component) {
            _inherits(LocalConsumerWrappedComponent, _Component);

            function LocalConsumerWrappedComponent() {
                var _ref2;

                for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
                    args[_key] = arguments[_key];
                }

                _classCallCheck(this, LocalConsumerWrappedComponent);

                var _this = _possibleConstructorReturn(this, (_ref2 = LocalConsumerWrappedComponent.__proto__ || Object.getPrototypeOf(LocalConsumerWrappedComponent)).call.apply(_ref2, [this].concat(args)));

                publicFn.forEach(function (fnName) {
                    _this[fnName] = function () {
                        var _this$child;

                        return _this.child && (_this$child = _this.child)[fnName].apply(_this$child, arguments);
                    };
                });
                return _this;
            }

            _createClass(LocalConsumerWrappedComponent, [{
                key: 'render',
                value: function render() {
                    var _this2 = this;

                    var _props = this.props,
                        locale = _props.locale,
                        rest = _objectWithoutProperties(_props, ['locale']);

                    var context = this.context.UC_FE_LOCALE || {};
                    return _react2.default.createElement(Child, _extends({
                        ref: function ref(_ref3) {
                            return _this2.child = _ref3;
                        },
                        locale: Object.assign({}, defaultLocale, context[localeName], requireRuntimeLocale ? (0, _runtime.getRuntimeLocale)()[localeName] : {}, locale)
                    }, rest));
                }
            }]);

            return LocalConsumerWrappedComponent;
        }(_react.Component), _class.propTypes = Object.assign({}, Child.PropTypes, {
            locale: _propTypes2.default.object
        }), _class.defaultProps = Object.assign({}, Child.defaultProps, {
            locale: {}
        }), _class.contextTypes = {
            UC_FE_LOCALE: _propTypes2.default.object
        }, _temp);

        return LocalConsumerWrappedComponent;
    };
};

exports.default = localeConsumerDecorator;
