// @ucloud-fe/react-components/lib/components/Checkbox/Group.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _store = require('../../decorators/selectableWithStore/store');

var _store2 = _interopRequireDefault(_store);

var _uncontrolled = require('../../decorators/uncontrolled');

var _uncontrolled2 = _interopRequireDefault(_uncontrolled);

var _Checkbox = require('./Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Group = (_dec = (0, _uncontrolled2.default)({ valueName: 'value' }), _dec2 = (0, _store2.default)({ StoreContext: _Checkbox.StoreContext }), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_Component) {
    _inherits(Group, _Component);

    function Group() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Group);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Group.__proto__ || Object.getPrototypeOf(Group)).call.apply(_ref, [this].concat(args))), _this), _this.renderOptions = function () {
            var _this$props = _this.props,
                options = _this$props.options,
                children = _this$props.children;

            if (options) {
                return options.map(function (option) {
                    var label = option.label,
                        restOptionProps = _objectWithoutProperties(option, ['label']);

                    return _react2.default.createElement(
                        _Checkbox2.default,
                        _extends({ key: option.value }, restOptionProps),
                        label !== undefined ? label : option.value
                    );
                });
            } else {
                return children;
            }
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Group, [{
        key: 'render',
        value: function render() {
            /* eslint-disable no-unused-vars */
            var _props = this.props,
                value = _props.value,
                defaultValue = _props.defaultValue,
                onChange = _props.onChange,
                options = _props.options,
                children = _props.children,
                selectable = _props.selectable,
                multiple = _props.multiple,
                disabled = _props.disabled,
                size = _props.size,
                styleType = _props.styleType,
                rest = _objectWithoutProperties(_props, ['value', 'defaultValue', 'onChange', 'options', 'children', 'selectable', 'multiple', 'disabled', 'size', 'styleType']);
            /* eslint-enable no-unused-vars */


            return _react2.default.createElement(
                _Checkbox.CheckboxContext.Provider,
                { value: Object.assign({}, (0, _pick3.default)(this.props, ['disabled', 'size', 'styleType'])) },
                _react2.default.createElement(
                    _style.CheckboxGroupWrap,
                    rest,
                    this.renderOptions()
                )
            );
        }
    }]);

    return Group;
}(_react.Component), _class2.propTypes = {
    /** 当前值，controlled */
    value: _propTypes2.default.array,
    /** 默认值，uncontrolled */
    defaultValue: _propTypes2.default.array,
    /** 修改时的回调 */
    onChange: _propTypes2.default.func,
    /** 快速定义选项 */
    options: _propTypes2.default.array,
    /** 禁用 */
    disabled: _propTypes2.default.bool,
    /** 尺寸 */
    size: _propTypes2.default.oneOf(_Checkbox2.default.Size),
    /** 样式风格 */
    styleType: _propTypes2.default.oneOf(_Checkbox2.default.StyleType),
    /** @ignore */
    children: _propTypes2.default.node,
    /** @ignore */
    selectable: _propTypes2.default.bool,
    /** @ignore */
    multiple: _propTypes2.default.bool
}, _class2.defaultProps = {
    defaultValue: [],
    onChange: function onChange() {},
    multiple: true,
    selectable: true
}, _temp2)) || _class) || _class);
exports.default = Group;
