// @ucloud-fe/react-components/lib/components/PopConfirm/PopConfirm.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Tooltip = require('../../components/Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _Button = require('../../components/Button');

var _Button2 = _interopRequireDefault(_Button);

var _SvgIcon = require('../../components/SvgIcon');

var _SvgIcon2 = _interopRequireDefault(_SvgIcon);

var _Combine = require('../../components/Combine');

var _Combine2 = _interopRequireDefault(_Combine);

var _localeConsumerDecorator = require('../../components/LocaleProvider/localeConsumerDecorator');

var _localeConsumerDecorator2 = _interopRequireDefault(_localeConsumerDecorator);

var _uncontrolled = require('../../decorators/uncontrolled');

var _uncontrolled2 = _interopRequireDefault(_uncontrolled);

var _style = require('./style');

var _zh_CN = require('./locale/zh_CN');

var _zh_CN2 = _interopRequireDefault(_zh_CN);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var noop = function noop() {};

var PopConfirm = (_dec = (0, _localeConsumerDecorator2.default)({ defaultLocale: _zh_CN2.default, localeName: 'PopConfirm' }), _dec2 = (0, _uncontrolled2.default)({ valueName: 'visible', onChangeName: 'onVisibleChange' }), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_Component) {
    _inherits(PopConfirm, _Component);

    function PopConfirm() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, PopConfirm);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = PopConfirm.__proto__ || Object.getPrototypeOf(PopConfirm)).call.apply(_ref, [this].concat(args))), _this), _this.onConfirm = function () {
            var _this$props = _this.props,
                onVisibleChange = _this$props.onVisibleChange,
                onConfirm = _this$props.onConfirm;

            onVisibleChange(false);
            onConfirm();
        }, _this.onCancel = function () {
            var _this$props2 = _this.props,
                onVisibleChange = _this$props2.onVisibleChange,
                onCancel = _this$props2.onCancel;

            onVisibleChange(false);
            onCancel();
        }, _this.renderPopup = function () {
            var _this$props3 = _this.props,
                locale = _this$props3.locale,
                popup = _this$props3.popup;

            return _react2.default.createElement(
                _style.PopupWrap,
                null,
                _react2.default.createElement(
                    _style.IconWrap,
                    null,
                    _react2.default.createElement(_SvgIcon2.default, { size: '20px', type: 'circleExclamation' })
                ),
                _react2.default.createElement(
                    _style.ContentWrap,
                    null,
                    popup
                ),
                _react2.default.createElement(
                    _style.FooterWrap,
                    null,
                    _react2.default.createElement(
                        _Combine2.default,
                        { sharedProps: { size: 'sm' } },
                        _react2.default.createElement(
                            _Button2.default,
                            { onClick: _this.onCancel },
                            locale.cancel
                        ),
                        _react2.default.createElement(
                            _Button2.default,
                            { onClick: _this.onConfirm, styleType: 'primary' },
                            locale.confirm
                        )
                    )
                )
            );
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(PopConfirm, [{
        key: 'render',
        value: function render() {
            // eslint-disable-next-line no-unused-vars
            var _props = this.props,
                popup = _props.popup,
                defaultVisible = _props.defaultVisible,
                visible = _props.visible,
                locale = _props.locale,
                rest = _objectWithoutProperties(_props, ['popup', 'defaultVisible', 'visible', 'locale']);

            return _react2.default.createElement(_Tooltip2.default, _extends({
                trigger: ['click'],
                popup: this.renderPopup(),
                customStyle: { popupWrapperPadding: '0px' }
            }, rest, {
                visible: visible
            }));
        }
    }]);

    return PopConfirm;
}(_react.Component), _class2.propTypes = {
    /** @ignore */
    popup: _propTypes2.default.node,
    /** @ignore */
    defaultVisible: _propTypes2.default.bool,
    /** @ignore */
    visible: _propTypes2.default.bool,
    /** @ignore */
    onVisibleChange: _propTypes2.default.func,
    /** @ignore */
    locale: _propTypes2.default.object,
    /** 确认按钮回调 */
    onConfirm: _propTypes2.default.func,
    /** 取消按钮回调 */
    onCancel: _propTypes2.default.func
}, _class2.defaultProps = {
    onVisibleChange: noop,
    onConfirm: noop,
    onCancel: noop
}, _temp2)) || _class) || _class);
exports.default = PopConfirm;
