// @ucloud-fe/react-components/lib/components/Popover/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.PopoverWrap = exports.animationPrefixCls = exports.prefixCls = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _rcTrigger = require('rc-trigger');

var _rcTrigger2 = _interopRequireDefault(_rcTrigger);

var _config = require('../../../config.js');

var _config2 = _interopRequireDefault(_config);

var _animation = require('../../../style/animation');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _prefixCls = _config2.default.prefixCls;
var prefixCls = exports.prefixCls = _prefixCls + '-popover';
var animationPrefixCls = exports.animationPrefixCls = prefixCls + '-animation';

var animationDuration = '0.1s';

var RcTriggerWrap = (_temp = _class = function (_Component) {
    _inherits(RcTriggerWrap, _Component);

    function RcTriggerWrap() {
        _classCallCheck(this, RcTriggerWrap);

        return _possibleConstructorReturn(this, (RcTriggerWrap.__proto__ || Object.getPrototypeOf(RcTriggerWrap)).apply(this, arguments));
    }

    _createClass(RcTriggerWrap, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                className = _props.className,
                popupClassName = _props.popupClassName,
                trueClassName = _props.trueClassName,
                triggerRef = _props.triggerRef,
                rest = _objectWithoutProperties(_props, ['className', 'popupClassName', 'trueClassName', 'triggerRef']);

            return _react2.default.createElement(_rcTrigger2.default, _extends({
                ref: triggerRef,
                className: trueClassName,
                popupClassName: (0, _classnames2.default)(className, popupClassName)
            }, rest));
        }
    }]);

    return RcTriggerWrap;
}(_react.Component), _class.propTypes = {
    className: _propTypes2.default.string,
    popupClassName: _propTypes2.default.string,
    trueClassName: _propTypes2.default.string,
    triggerRef: _propTypes2.default.func
}, _temp);
var PopoverWrap = exports.PopoverWrap = (0, _styledComponents2.default)(RcTriggerWrap).withConfig({
    displayName: 'style__PopoverWrap',
    componentId: 'sc-1yqouou-0'
})(['&.', '{position:absolute;left:-9999px;top:-9999px;z-index:1030;display:block;&-hidden{display:none;}}&.', '-animation{&-fade{&-enter,&-appear,&-leave{animation-duration:', ';animation-fill-mode:both;}&-enter,&-appear{animation-name:', ';}&-leave{animation-name:', ';}}&-bounce{&-enter,&-appear,&-leave{animation-duration:', ';animation-fill-mode:both;}&-enter,&-appear{animation-name:', ';}&-leave{animation-name:', ';}}&-zoom{&-enter,&-appear,&-leave{animation-duration:', ';animation-fill-mode:both;}&-enter,&-appear{animation-name:', ';}&-leave{animation-name:', ';}}&-slide-up{&-enter,&-appear{animation-duration:', ';animation-fill-mode:both;transform-origin:0 0;display:block !important;opacity:0;animation-timing-function:cubic-bezier(0.08,0.82,0.17,1);animation-play-state:paused;}&-leave{animation-duration:', ';animation-fill-mode:both;transform-origin:0 0;display:block !important;opacity:1;animation-timing-function:cubic-bezier(0.6,0.04,0.98,0.34);animation-play-state:paused;}&-enter-active.', '-placement-bottomLeft,&-appear-active.', '-placement-bottomLeft,&-enter-active.', '-placement-bottomCenter,&-appear-active.', '-placement-bottomCenter,&-enter-active.', '-placement-bottomRight,&-appear-active.', '-placement-bottomRight{animation-name:', ';animation-play-state:running;}&-enter-active.', '-placement-topLeft,&-appear-active.', '-placement-topLeft,&-enter-active.', '-placement-topCenter,&-appear-active.', '-placement-topCenter,&-enter-active.', '-placement-topRight,&-appear-active.', '-placement-topRight{animation-name:', ';animation-play-state:running;}&-leave-active.', '-placement-bottomLeft,&-leave-active.', '-placement-bottomCenter,&-leave-active.', '-placement-bottomRight{animation-name:', ';animation-play-state:running;}&-leave-active.', '-placement-topLeft,&-leave-active.', '-placement-topCenter,&-leave-active.', '-placement-topRight{animation-name:', ';animation-play-state:running;}}}'], prefixCls, prefixCls, animationDuration, _animation.fadeIn, _animation.fadeOut, animationDuration, _animation.bounceIn, _animation.bounceOut, animationDuration, _animation.zoomIn, _animation.zoomOut, animationDuration, animationDuration, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, _animation.slideUpIn, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, _animation.slideDownIn, prefixCls, prefixCls, prefixCls, _animation.slideUpOut, prefixCls, prefixCls, prefixCls, _animation.slideDownOut);
