// @ucloud-fe/react-components/lib/components/Badge/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.DotBadge = exports.BaseBadge = exports.BadgeWrap = exports.Wrap = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _prefixCls = _config2.default.prefixCls;

var prefixCls = _prefixCls + '-badge';

var Wrap = exports.Wrap = _styledComponents2.default.div.attrs({
    className: prefixCls
}).withConfig({
    displayName: 'style__Wrap',
    componentId: 'sc-1deoyov-0'
})(['position:relative;display:inline-block;']);

var BadgeWrap = exports.BadgeWrap = _styledComponents2.default.div.attrs({
    className: prefixCls + '-badge-wrap'
}).withConfig({
    displayName: 'style__BadgeWrap',
    componentId: 'sc-1deoyov-1'
})(['position:absolute;']);

var BaseBadge = exports.BaseBadge = _styledComponents2.default.span.attrs({
    className: prefixCls + '-badge'
}).withConfig({
    displayName: 'style__BaseBadge',
    componentId: 'sc-1deoyov-2'
})(['display:inline-block;min-height:20px;line-height:20px;min-width:10px;padding:0 5px;border-radius:10px;text-align:center;background:red;color:white;']);

var DotBadge = exports.DotBadge = (0, _styledComponents2.default)(BaseBadge).attrs({
    className: prefixCls + '-badge-dot'
}).withConfig({
    displayName: 'style__DotBadge',
    componentId: 'sc-1deoyov-3'
})(['width:10px;height:10px;padding:0;border-radius:5px;min-width:0;min-height:0;']);
