// @ucloud-fe/react-components/lib/components/Modal/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ModalWrap = exports.prefixCls = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp;

var _templateObject = _taggedTemplateLiteral(['\n    .', ' {\n        &-mask {\n            position: fixed;\n            top: 0;\n            right: 0;\n            left: 0;\n            bottom: 0;\n            height: 100%;\n            z-index: 1010;\n            &-hidden {\n                display: none;\n            }\n        }\n    }\n'], ['\n    .', ' {\n        &-mask {\n            position: fixed;\n            top: 0;\n            right: 0;\n            left: 0;\n            bottom: 0;\n            height: 100%;\n            z-index: 1010;\n            &-hidden {\n                display: none;\n            }\n        }\n    }\n']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _rcDialog = require('rc-dialog');

var _rcDialog2 = _interopRequireDefault(_rcDialog);

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

require('../../../style/globalAnimation');

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var _prefixCls = _config2.default.prefixCls;
var prefixCls = exports.prefixCls = _prefixCls + '-modal';

(0, _styledComponents.injectGlobal)(_templateObject, prefixCls);

var RcDialogWrap = (_temp = _class = function (_Component) {
    _inherits(RcDialogWrap, _Component);

    function RcDialogWrap() {
        _classCallCheck(this, RcDialogWrap);

        return _possibleConstructorReturn(this, (RcDialogWrap.__proto__ || Object.getPrototypeOf(RcDialogWrap)).apply(this, arguments));
    }

    _createClass(RcDialogWrap, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                className = _props.className,
                wrapClassName = _props.wrapClassName,
                trueClassName = _props.trueClassName,
                rest = _objectWithoutProperties(_props, ['className', 'wrapClassName', 'trueClassName']);

            return _react2.default.createElement(_rcDialog2.default, _extends({ className: trueClassName, wrapClassName: (0, _classnames2.default)(className, wrapClassName) }, rest));
        }
    }]);

    return RcDialogWrap;
}(_react.Component), _class.propTypes = {
    className: _propTypes2.default.string,
    wrapClassName: _propTypes2.default.string,
    trueClassName: _propTypes2.default.string
}, _temp);
var ModalWrap = exports.ModalWrap = (0, _styledComponents2.default)(RcDialogWrap)(function (_ref) {
    var _ref$theme = _ref.theme,
        DT = _ref$theme.designTokens,
        fontSize = _ref$theme.fontSize,
        mask = _ref.mask;
    return (0, _styledComponents.css)(['position:fixed;overflow:auto;top:0;right:0;bottom:0;left:0;z-index:1010;-webkit-overflow-scrolling:touch;outline:0;background:', ';.', '{padding:0;line-height:1;font-size:', ';color:', ';position:relative;width:340px;margin:0 auto 100px;top:15%;border-radius:4px;background:', ';box-shadow:', ';&-content{position:relative;border:none;border-radius:6px 6px;background-clip:padding-box;}&-close{position:absolute;z-index:1;right:0;top:0;padding:16px;margin:0;width:20px;height:20px;font-size:20px;cursor:pointer;color:', ';}&-close-x{display:none;}&-title-content{position:relative;padding:16px 52px 16px 16px;min-height:20px;color:', ';border-bottom:1px solid ', ';border-radius:4px 4px 0 0;font-size:16px;line-height:20px;}&-footer{padding:15px 16px;border-top:1px solid ', ';border-radius:0 0 4px 4px;line-height:1;text-align:right;}&-body{min-height:60px;}}'], mask && DT.T_MODAL_COLOR_LAYER_DEFAULT, prefixCls, fontSize, DT.T_COLOR_TEXT_DEFAULT_DARK, DT.T_COLOR_BG_DEFAULT_NORMAL, DT.T_MODAL_SHADOW_DEFAULT, DT.T_COLOR_TEXT_DEFAULT_LIGHT, DT.T_COLOR_TEXT_DEFAULT_DARK, DT.T_COLOR_LINE_DEFAULT_DARK, DT.T_COLOR_LINE_DEFAULT_DARK);
});
(0, _addDefaultThemeProps2.default)(ModalWrap);
