// @ucloud-fe/react-components/lib/components/Badge/Badge.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _keys2 = require('lodash/keys');

var _keys3 = _interopRequireDefault(_keys2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _rcAlign = require('rc-align');

var _rcAlign2 = _interopRequireDefault(_rcAlign);

var _style = require('./style');

var _placements = require('./placements');

var _placements2 = _interopRequireDefault(_placements);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Placement = (0, _keys3.default)(_placements2.default);

var Badge = (_temp2 = _class = function (_Component) {
    _inherits(Badge, _Component);

    function Badge() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Badge);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Badge.__proto__ || Object.getPrototypeOf(Badge)).call.apply(_ref, [this].concat(args))), _this), _this.renderBadge = function () {
            var _this$props = _this.props,
                _v = _this$props.value,
                maxValue = _this$props.maxValue,
                dot = _this$props.dot,
                badgeStyle = _this$props.badgeStyle;

            var isNumber = /^\d+\.?\d*$/.test(_v);
            var value = isNumber ? +_v : _v;

            var content = void 0;
            if (dot) {
                content = null;
            } else if (isNumber && value > maxValue) {
                content = maxValue + '+';
            } else {
                content = value;
            }
            var Comp = dot ? _style.DotBadge : _style.BaseBadge;
            return _react2.default.createElement(
                Comp,
                { style: badgeStyle },
                content
            );
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Badge, [{
        key: 'render',
        value: function render() {
            var _this2 = this;

            /* eslint-disable no-unused-vars */
            var _props = this.props,
                value = _props.value,
                maxValue = _props.maxValue,
                dot = _props.dot,
                placement = _props.placement,
                hideWhenZero = _props.hideWhenZero,
                children = _props.children,
                badgeStyle = _props.badgeStyle,
                innerRef = _props.innerRef,
                rest = _objectWithoutProperties(_props, ['value', 'maxValue', 'dot', 'placement', 'hideWhenZero', 'children', 'badgeStyle', 'innerRef']);
            /* eslint-enable no-unused-vars */


            var badge = this.renderBadge();
            if (!children) {
                return _react2.default.createElement(
                    _style.Wrap,
                    rest,
                    badge
                );
            }
            /* eslint-disable react/no-find-dom-node */
            return _react2.default.createElement(
                _style.Wrap,
                rest,
                children,
                hideWhenZero && value === 0 ? null : _react2.default.createElement(
                    _rcAlign2.default,
                    { ref: innerRef, target: function target() {
                            return _reactDom2.default.findDOMNode(_this2);
                        }, align: _placements2.default[placement] },
                    _react2.default.createElement(
                        _style.BadgeWrap,
                        null,
                        badge
                    )
                )
            );
            /* eslint-enable react/no-find-dom-node */
        }
    }]);

    return Badge;
}(_react.Component), _class.propTypes = {
    /** 显示内容 */
    value: _propTypes2.default.node,
    /** 为数字时能显示的最大值，超过将显示最大值+ */
    maxValue: _propTypes2.default.number,
    /** 显示为点状 */
    dot: _propTypes2.default.bool,
    /** 定位 */
    placement: _propTypes2.default.oneOf(Placement),
    /** 为0时是否隐藏 */
    hideWhenZero: _propTypes2.default.bool,
    /** badge的样式 */
    badgeStyle: _propTypes2.default.object,
    /** @ignore */
    children: _propTypes2.default.node,
    /** @ignore */
    innerRef: _propTypes2.default.any
}, _class.defaultProps = {
    maxValue: 99,
    placement: Placement[0]
}, _temp2);
exports.default = Badge;


Badge.Placement = Placement;
