// @ucloud-fe/react-components/lib/components/Popover/placements.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
var autoAdjustOverflow = {
    adjustX: 1,
    adjustY: 1
};

var targetOffset = [0, 0];

var getPlacements = exports.getPlacements = function getPlacements() {
    var offset = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 5;

    return {
        topLeft: {
            points: ['bl', 'tl'],
            overflow: autoAdjustOverflow,
            offset: [0, -offset],
            targetOffset: targetOffset
        },
        top: {
            points: ['bc', 'tc'],
            overflow: autoAdjustOverflow,
            offset: [0, -offset],
            targetOffset: targetOffset
        },
        topRight: {
            points: ['br', 'tr'],
            overflow: autoAdjustOverflow,
            offset: [0, -offset],
            targetOffset: targetOffset
        },
        bottomLeft: {
            points: ['tl', 'bl'],
            overflow: autoAdjustOverflow,
            offset: [0, offset],
            targetOffset: targetOffset
        },
        bottom: {
            points: ['tc', 'bc'],
            overflow: autoAdjustOverflow,
            offset: [0, offset],
            targetOffset: targetOffset
        },
        bottomRight: {
            points: ['tr', 'br'],
            overflow: autoAdjustOverflow,
            offset: [0, offset],
            targetOffset: targetOffset
        },
        leftTop: {
            points: ['tr', 'tl'],
            overflow: autoAdjustOverflow,
            offset: [-offset, 0],
            targetOffset: targetOffset
        },
        left: {
            points: ['cr', 'cl'],
            overflow: autoAdjustOverflow,
            offset: [-offset, 0],
            targetOffset: targetOffset
        },
        leftBottom: {
            points: ['br', 'bl'],
            overflow: autoAdjustOverflow,
            offset: [-offset, 0],
            targetOffset: targetOffset
        },
        rightTop: {
            points: ['tl', 'tr'],
            overflow: autoAdjustOverflow,
            offset: [offset, 0],
            targetOffset: targetOffset
        },
        right: {
            points: ['cl', 'cr'],
            overflow: autoAdjustOverflow,
            offset: [offset, 0],
            targetOffset: targetOffset
        },
        rightBottom: {
            points: ['bl', 'br'],
            overflow: autoAdjustOverflow,
            offset: [offset, 0],
            targetOffset: targetOffset
        }
    };
};

exports.default = getPlacements();
