// @ucloud-fe/react-components/lib/components/Radio/Radio.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.RadioContext = exports.StoreContext = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createReactContext = require('create-react-context');

var _createReactContext2 = _interopRequireDefault(_createReactContext);

var _item = require('../../decorators/selectableWithStore/item');

var _item2 = _interopRequireDefault(_item);

var _uncontrolled = require('../../decorators/uncontrolled');

var _uncontrolled2 = _interopRequireDefault(_uncontrolled);

var _SvgIcon = require('../../components/SvgIcon');

var _SvgIcon2 = _interopRequireDefault(_SvgIcon);

var _style = require('./style');

var _Card = require('./Card');

var _Card2 = _interopRequireDefault(_Card);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var StoreContext = exports.StoreContext = (0, _createReactContext2.default)();
var RadioContext = exports.RadioContext = (0, _createReactContext2.default)();

var Size = ['sm', 'md', 'lg'];
var StyleType = ['default', 'button', 'tag', 'card', 'text'];

var Radio = (_dec = (0, _uncontrolled2.default)({ valueName: 'checked' }), _dec2 = (0, _item2.default)({ StoreContext: StoreContext }), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_Component) {
    _inherits(Radio, _Component);

    function Radio() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Radio);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Radio.__proto__ || Object.getPrototypeOf(Radio)).call.apply(_ref, [this].concat(args))), _this), _this.onClick = function (props, e) {
            var onChange = props.onChange,
                onClick = props.onClick,
                disabled = props.disabled,
                checked = props.checked;


            if (disabled) return;

            onClick(e);

            if (!checked) onChange(true);
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Radio, [{
        key: 'renderRadio',
        value: function renderRadio(props) {
            var _this2 = this;

            /* eslint-disable no-unused-vars */
            var children = props.children,
                checked = props.checked,
                defaultChecked = props.defaultChecked,
                value = props.value,
                onChange = props.onChange,
                onClick = props.onClick,
                multiple = props.multiple,
                disabled = props.disabled,
                title = props.title,
                rest = _objectWithoutProperties(props, ['children', 'checked', 'defaultChecked', 'value', 'onChange', 'onClick', 'multiple', 'disabled', 'title']);
            /* eslint-enable no-unused-vars */

            return _react2.default.createElement(
                _style.RadioWrap,
                _extends({
                    checked: checked,
                    disabled: disabled
                }, rest, {
                    onClick: function onClick() {
                        for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
                            args[_key2] = arguments[_key2];
                        }

                        return _this2.onClick.apply(_this2, [props].concat(args));
                    }
                }),
                _react2.default.createElement(
                    'span',
                    { className: _style.iconWrapCls },
                    _react2.default.createElement(_SvgIcon2.default, { className: _style.iconCls, type: 'circle', size: '14px' })
                ),
                children != null && _react2.default.createElement(
                    'span',
                    { className: _style.contentCls },
                    children
                )
            );
        }
    }, {
        key: 'renderRadioButton',
        value: function renderRadioButton(props) {
            var _this3 = this;

            /* eslint-disable no-unused-vars */
            var title = props.title,
                disabledLabel = props.disabledLabel,
                onChange = props.onChange,
                rest = _objectWithoutProperties(props, ['title', 'disabledLabel', 'onChange']);
            /* eslint-enable no-unused-vars */

            return _react2.default.createElement(_style.RadioButtonWrap, _extends({}, rest, { onClick: function onClick() {
                    for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
                        args[_key3] = arguments[_key3];
                    }

                    return _this3.onClick.apply(_this3, [props].concat(args));
                } }));
        }
    }, {
        key: 'renderRadioTag',
        value: function renderRadioTag(props) {
            var _this4 = this;

            /* eslint-disable no-unused-vars */
            var title = props.title,
                disabledLabel = props.disabledLabel,
                onChange = props.onChange,
                rest = _objectWithoutProperties(props, ['title', 'disabledLabel', 'onChange']);
            /* eslint-enable no-unused-vars */

            return _react2.default.createElement(_style.RadioTagWrap, _extends({}, rest, { onClick: function onClick() {
                    for (var _len4 = arguments.length, args = Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
                        args[_key4] = arguments[_key4];
                    }

                    return _this4.onClick.apply(_this4, [props].concat(args));
                } }));
        }
    }, {
        key: 'renderRadioCard',
        value: function renderRadioCard(props) {
            var _this5 = this;

            /* eslint-disable no-unused-vars */
            var onChange = props.onChange,
                rest = _objectWithoutProperties(props, ['onChange']);
            /* eslint-enable no-unused-vars */


            return _react2.default.createElement(_Card2.default, _extends({}, rest, { onClick: function onClick() {
                    for (var _len5 = arguments.length, args = Array(_len5), _key5 = 0; _key5 < _len5; _key5++) {
                        args[_key5] = arguments[_key5];
                    }

                    return _this5.onClick.apply(_this5, [props].concat(args));
                } }));
        }
    }, {
        key: 'renderRadioText',
        value: function renderRadioText(props) {
            var _this6 = this;

            /* eslint-disable no-unused-vars */
            var title = props.title,
                disabledLabel = props.disabledLabel,
                children = props.children,
                onChange = props.onChange,
                rest = _objectWithoutProperties(props, ['title', 'disabledLabel', 'children', 'onChange']);
            /* eslint-enable no-unused-vars */


            return _react2.default.createElement(
                _style.RadioTextWrap,
                _extends({}, rest, { onClick: function onClick() {
                        for (var _len6 = arguments.length, args = Array(_len6), _key6 = 0; _key6 < _len6; _key6++) {
                            args[_key6] = arguments[_key6];
                        }

                        return _this6.onClick.apply(_this6, [props].concat(args));
                    } }),
                _react2.default.createElement(
                    'span',
                    null,
                    _react2.default.createElement(
                        'span',
                        null,
                        children
                    )
                )
            );
        }
    }, {
        key: 'render',
        value: function render() {
            var _this7 = this;

            return _react2.default.createElement(
                RadioContext.Consumer,
                null,
                function (context) {
                    /* eslint-disable no-unused-vars */
                    var _props = _this7.props,
                        defaultChecked = _props.defaultChecked,
                        value = _props.value,
                        multiple = _props.multiple,
                        restProps = _objectWithoutProperties(_props, ['defaultChecked', 'value', 'multiple']);
                    /* eslint-enable no-unused-vars */


                    var props = Object.assign({
                        size: 'md',
                        styleType: 'default'
                    }, context, restProps);

                    var styleType = props.styleType;

                    switch (styleType) {
                        case 'button':
                            return _this7.renderRadioButton(props);
                        case 'tag':
                            return _this7.renderRadioTag(props);
                        case 'card':
                            return _this7.renderRadioCard(props);
                        case 'text':
                            return _this7.renderRadioText(props);
                        default:
                            return _this7.renderRadio(props);
                    }
                }
            );
        }
    }]);

    return Radio;
}(_react.Component), _class2.propTypes = {
    /** @ignore */
    children: _propTypes2.default.node,
    /** 是否选中 */
    checked: _propTypes2.default.bool,
    /** 默认是否选中 */
    defaultChecked: _propTypes2.default.bool,
    /** 是否禁用 */
    disabled: _propTypes2.default.bool,
    /** 点选时的回调 */
    onChange: _propTypes2.default.func,
    /** @ignore */
    onClick: _propTypes2.default.func,
    /** radio的值 */
    value: _propTypes2.default.any,
    /** 样式风格 */
    styleType: _propTypes2.default.oneOf(StyleType),
    /** 尺寸，styleType为card时无效 */
    size: _propTypes2.default.oneOf(Size),
    /** 标题，styleType为card时使用 */
    title: _propTypes2.default.node,
    /** 禁用标识，styleType为card时使用 */
    disabledLabel: _propTypes2.default.node,
    /** @ignore */
    multiple: _propTypes2.default.bool
}, _class2.defaultProps = {
    onChange: function onChange() {},
    onClick: function onClick() {}
}, _temp2)) || _class) || _class);


Radio.Size = Size;
Radio.StyleType = StyleType;

exports.default = Radio;
