// @ucloud-fe/react-components/lib/components/Combine/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.CombineWrap = exports.controllerPrefix = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _config = require('../../../config.js');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _prefixCls = _config2.default.prefixCls;

var prefixCls = _prefixCls + '-combine';

var controllerPrefix = exports.controllerPrefix = prefixCls + '-controller';

var spacingMap = {
    compact: '-1px',
    sm: '8px',
    md: '12px',
    lg: '12px'
};

var CombineWrap = exports.CombineWrap = _styledComponents2.default.div.attrs({
    className: prefixCls
})(function (_ref) {
    var spacing = _ref.spacing;
    return (0, _styledComponents.css)(['position:relative;> .', '{vertical-align:middle;display:inline-block;}> .', '+.', '{margin-left:', ';&:focus{z-index:2;}&:hover{z-index:3;}}'], controllerPrefix, controllerPrefix, controllerPrefix, spacingMap[spacing] || spacing);
});
