// @ucloud-fe/react-components/lib/components/Form/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.FormWrap = exports.GroupTitle = exports.GroupWrap = exports.ControllerWrap = exports.LabelWrap = exports.ItemWrap = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _Grid = require('../../../components/Grid');

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _prefixCls = _config2.default.prefixCls;

var prefixCls = _prefixCls + '-form';

var ItemWrap = exports.ItemWrap = (0, _styledComponents2.default)(_Grid.Row).attrs({
    className: prefixCls + '-item'
}).withConfig({
    displayName: 'style__ItemWrap',
    componentId: 'sc-15nkpst-0'
})(['']);

var LabelWrap = exports.LabelWrap = (0, _styledComponents2.default)(_Grid.Col).attrs({
    className: prefixCls + '-label'
}).withConfig({
    displayName: 'style__LabelWrap',
    componentId: 'sc-15nkpst-1'
})(['padding-top:4px;line-height:20px;word-break:break-all;']);

var ControllerWrap = exports.ControllerWrap = (0, _styledComponents2.default)(_Grid.Col).attrs({
    className: prefixCls + '-controller'
}).withConfig({
    displayName: 'style__ControllerWrap',
    componentId: 'sc-15nkpst-2'
})(['']);

var GroupWrap = exports.GroupWrap = _styledComponents2.default.div.attrs({
    className: prefixCls + '-group'
}).withConfig({
    displayName: 'style__GroupWrap',
    componentId: 'sc-15nkpst-3'
})(['& + &{margin-top:24px;}']);

var GroupTitle = exports.GroupTitle = _styledComponents2.default.div.attrs({
    className: prefixCls + '-group-title'
}).withConfig({
    displayName: 'style__GroupTitle',
    componentId: 'sc-15nkpst-4'
})(['font-weight:bold;line-height:22px;padding:12px 0;margin-bottom:24px;', ';'], function (_ref) {
    var _ref$theme = _ref.theme,
        colorList = _ref$theme.colorList,
        colorMap = _ref$theme.colorMap,
        titleFontSize = _ref$theme.titleFontSize;
    return (0, _styledComponents.css)(['font-size:', ';border-bottom:1px solid ', ';color:', ';'], titleFontSize, colorMap.default.border, colorList.title);
});

var FormWrap = exports.FormWrap = _styledComponents2.default.form.attrs({
    className: prefixCls
})(function (_ref2) {
    var size = _ref2.size;
    return (0, _styledComponents.css)(['', '{margin-bottom:16px;&:last-child{margin-bottom:0;}}', ';'], ItemWrap, size === 'lg' && (0, _styledComponents.css)(['', '{margin-bottom:24px;}', '{padding-top:6px;}'], ItemWrap, LabelWrap));
});

(0, _addDefaultThemeProps2.default)(GroupTitle);
