// @ucloud-fe/react-components/lib/components/Collapse/Collapse.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.StoreContext = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createReactContext = require('create-react-context');

var _createReactContext2 = _interopRequireDefault(_createReactContext);

var _store = require('../../decorators/selectableWithStore/store');

var _store2 = _interopRequireDefault(_store);

var _uncontrolled = require('../../decorators/uncontrolled');

var _uncontrolled2 = _interopRequireDefault(_uncontrolled);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var StoreContext = exports.StoreContext = (0, _createReactContext2.default)();

var Collapse = (_dec = (0, _uncontrolled2.default)({ valueName: 'openKeys' }), _dec2 = (0, _store2.default)({
    valueName: 'openKeys',
    childValueName: 'panelKey',
    StoreContext: StoreContext
}), _dec(_class = _dec2(_class = (_temp = _class2 = function (_Component) {
    _inherits(Collapse, _Component);

    function Collapse() {
        _classCallCheck(this, Collapse);

        return _possibleConstructorReturn(this, (Collapse.__proto__ || Object.getPrototypeOf(Collapse)).apply(this, arguments));
    }

    _createClass(Collapse, [{
        key: 'render',
        value: function render() {
            // eslint-disable-next-line no-unused-vars
            var _props = this.props,
                openKeys = _props.openKeys,
                defaultOpenKeys = _props.defaultOpenKeys,
                multiple = _props.multiple,
                selectable = _props.selectable,
                onChange = _props.onChange,
                theme = _props.theme,
                rest = _objectWithoutProperties(_props, ['openKeys', 'defaultOpenKeys', 'multiple', 'selectable', 'onChange', 'theme']);

            return _react2.default.createElement('div', rest);
        }
    }]);

    return Collapse;
}(_react.Component), _class2.propTypes = {
    /** 激活的面板，multiple时为数组值，controlled */
    openKeys: _propTypes2.default.any,
    /** 默认激活的面板，multiple时为数组值，uncontrolled */
    defaultOpenKeys: _propTypes2.default.array,
    /** 是否可以多个同时打开 */
    multiple: _propTypes2.default.bool,
    /** @ignore */
    selectable: _propTypes2.default.bool,
    /** @ignore */
    theme: _propTypes2.default.any,
    /** 变化回调 */
    onChange: _propTypes2.default.func
}, _class2.defaultProps = {
    selectable: true,
    multiple: true
}, _temp)) || _class) || _class);
exports.default = Collapse;
