// @ucloud-fe/react-components/lib/components/Drawer/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.DrawerWrap = exports.CloseHandlerWrapper = exports.prefixCls = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _rcDrawer = require('../../../libs/rc-drawer');

var _rcDrawer2 = _interopRequireDefault(_rcDrawer);

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _prefixCls = _config2.default.prefixCls;
var prefixCls = exports.prefixCls = _prefixCls + '-drawer';

var propsMixin = function propsMixin(_ref) {
    var zIndex = _ref.zIndex,
        show = _ref.show,
        DT = _ref.theme.designTokens,
        getContainer = _ref.getContainer;
    return (0, _styledComponents.css)(['background:', ';.', '-mask,.', '-content-wrapper{z-index:', ';position:', ';}&.', '-open.', '-left .', '-content-wrapper{box-shadow:', ';}&.', '-open.', '-right .', '-content-wrapper{box-shadow:', ';}&.', '-open.', '-top .', '-content-wrapper{box-shadow:', ';}&.', '-open.', '-bottom .', '-content-wrapper{box-shadow:', ';}', ';'], DT.T_COLOR_BG_CONTENT_1, prefixCls, prefixCls, zIndex, getContainer ? 'absolute' : 'fixed', prefixCls, prefixCls, prefixCls, DT.T_SHADOW_BLOCK_RIGHT_LG, prefixCls, prefixCls, prefixCls, DT.T_SHADOW_BLOCK_LEFT_LG, prefixCls, prefixCls, prefixCls, DT.T_SHADOW_BLOCK_BOTTOM_LG, prefixCls, prefixCls, prefixCls, DT.T_SHADOW_BLOCK_TOP_LG, show ? (0, _styledComponents.css)(['.', '-mask{top:0;left:0;right:0;bottom:0;}'], prefixCls) : (0, _styledComponents.css)(['']));
};

// eslint-disable-next-line react/prop-types,no-unused-vars
var CleanPropsRcDrawer = function CleanPropsRcDrawer(_ref2) {
    var show = _ref2.show,
        theme = _ref2.theme,
        zIndex = _ref2.zIndex,
        rest = _objectWithoutProperties(_ref2, ['show', 'theme', 'zIndex']);

    return _react2.default.createElement(_rcDrawer2.default, rest);
};

var closeHandlerPropsMixin = function closeHandlerPropsMixin(_ref3) {
    var DT = _ref3.theme.designTokens;
    return (0, _styledComponents.css)(['background:', ';box-shadow:', ';:hover{background:', ';box-shadow:', ';}'], DT.T_BUTTON_PRIMARY_COLOR_BG_DEFAULT, DT.T_SHADOW_BUTTON_PRIMARY, DT.T_BUTTON_PRIMARY_COLOR_BG_HOVER, DT.T_SHADOW_BUTTON_PRIMARY_HOVER);
};

var CloseHandlerWrapper = exports.CloseHandlerWrapper = _styledComponents2.default.span.withConfig({
    displayName: 'style__CloseHandlerWrapper',
    componentId: 'sc-192pngt-0'
})(function (_ref4) {
    var DT = _ref4.theme.designTokens;
    return (0, _styledComponents.css)(['display:inline-block;text-align:center;position:absolute;cursor:pointer;font-size:0px;transition:all 0.3s;', ';fill:', ';'], closeHandlerPropsMixin, DT.T_BUTTON_PRIMARY_COLOR_TEXT_DEFAULT);
});

/* stylelint-disable no-descending-specificity */
var DrawerWrap = exports.DrawerWrap = (0, _styledComponents2.default)(CleanPropsRcDrawer).attrs({
    prefixCls: prefixCls
})(function (_ref5) {
    var DT = _ref5.theme.designTokens;
    return (0, _styledComponents.css)(['outline:none;.', '-mask{background:', ';opacity:0;}&.', '-open{.', '-mask{opacity:1;}}.', '-content{height:100%;width:100%;overflow:auto;}.', '-content-wrapper{height:100%;width:100%;position:absolute;background:', ';}.', '-content-wrapper,.', '-mask{transition:all 0.1s;}&.', '-left,&.', '-right{', '{width:24px;height:48px;line-height:48px;top:110px;}}&.', '-left{.', '-content-wrapper{top:0;left:0;height:100%;}', '{right:-24px;border-radius:0px 4px 4px 0px;}}&.', '-right{.', '-content-wrapper{top:0;right:0;height:100%;}', '{left:-24px;border-radius:4px 0px 0px 4px;}}&.', '-top,&.', '-bottom{', '{width:48px;height:24px;line-height:24px;left:50%;margin-left:-24px;}}&.', '-top{.', '-content-wrapper{top:0;left:0;width:100%;}', '{bottom:-24px;border-radius:0px 0px 4px 4px;}}&.', '-bottom{.', '-content-wrapper{bottom:0;left:0;width:100%;}', '{top:-24px;border-radius:4px 4px 0px 0px;}}', ';'], prefixCls, DT.T_MODAL_COLOR_LAYER_DEFAULT, prefixCls, prefixCls, prefixCls, prefixCls, DT.T_COLOR_BG_DEFAULT_NORMAL, prefixCls, prefixCls, prefixCls, prefixCls, CloseHandlerWrapper, prefixCls, prefixCls, CloseHandlerWrapper, prefixCls, prefixCls, CloseHandlerWrapper, prefixCls, prefixCls, CloseHandlerWrapper, prefixCls, prefixCls, CloseHandlerWrapper, prefixCls, prefixCls, CloseHandlerWrapper, propsMixin);
});

(0, _addDefaultThemeProps2.default)(DrawerWrap, CloseHandlerWrapper);
