// @ucloud-fe/react-components/lib/components/Steps/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.StepsWrapper = exports.LinkWrapper = exports.StepWrapper = exports.RemarkWrapper = exports.TitleWrapper = exports.ContentWrapper = exports.StepCountWrapper = exports.IconWrapper = exports.Icon = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _SvgIcon = require('../../../components/SvgIcon');

var _SvgIcon2 = _interopRequireDefault(_SvgIcon);

var _style = require('../../../style');

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Icon = exports.Icon = (0, _styledComponents2.default)(_SvgIcon2.default).attrs({ size: '16px' }).withConfig({
    displayName: 'style__Icon',
    componentId: 'uqtuoq-0'
})(['']);

var statusMixin = function statusMixin(_ref) {
    var status = _ref.status,
        DT = _ref.theme.designTokens;

    switch (status) {
        case 'before':
            return (0, _styledComponents.css)(['background:', ';border-color:', ';color:', ';fill:', ';'], DT.T_COLOR_BG_PRIMARY_5, DT.T_COLOR_LINE_PRIMARY_DEFAULT, DT.T_COLOR_TEXT_PRIMARY_DEFAULT, DT.T_COLOR_TEXT_PRIMARY_DEFAULT);
        case 'after':
            return (0, _styledComponents.css)(['background:', ';border-color:', ';color:', ';fill:', ';'], DT.T_COLOR_BG_DISABLED_LIGHT, DT.T_COLOR_LINE_DISABLED_LIGHT, DT.T_COLOR_TEXT_DISABLED, DT.T_COLOR_TEXT_DISABLED);
        case 'current':
        case 'loading':
            return (0, _styledComponents.css)(['background:', ';border-color:', ';color:', ';fill:', ';'], DT.T_COLOR_BG_PRIMARY_1, DT.T_COLOR_BG_PRIMARY_1, DT.T_BUTTON_PRIMARY_COLOR_TEXT_DEFAULT, DT.T_BUTTON_PRIMARY_COLOR_TEXT_DEFAULT);
        case 'error':
            return (0, _styledComponents.css)(['background:', ';border-color:', ';color:', ';fill:', ';'], DT.T_COLOR_BG_ERROR_LIGHT, DT.T_COLOR_LINE_ERROR_DARK, DT.T_COLOR_TEXT_ERROR, DT.T_COLOR_TEXT_ERROR);
    }
};

var IconWrapper = exports.IconWrapper = _styledComponents2.default.span.withConfig({
    displayName: 'style__IconWrapper',
    componentId: 'uqtuoq-1'
})(['width:32px;height:32px;line-height:30px;text-align:center;box-sizing:border-box;border-radius:50%;border:1px solid;font-size:0;display:inline-block;vertical-align:top;transition:all 0.3s;', ';', ';'], statusMixin, function (_ref2) {
    var spin = _ref2.spin;
    return spin && _style.spinMixin;
});

var StepCountWrapper = exports.StepCountWrapper = _styledComponents2.default.span.withConfig({
    displayName: 'style__StepCountWrapper',
    componentId: 'uqtuoq-2'
})(['font-size:16px;']);

var ContentWrapper = exports.ContentWrapper = _styledComponents2.default.span.withConfig({
    displayName: 'style__ContentWrapper',
    componentId: 'uqtuoq-3'
})(['margin-left:12px;', ';'], _style.inlineBlockWithVerticalMixin);

var TitleWrapper = exports.TitleWrapper = _styledComponents2.default.span.withConfig({
    displayName: 'style__TitleWrapper',
    componentId: 'uqtuoq-4'
})(['font-size:14px;line-height:32px;min-height:32px;display:block;transition:all 0.3s;']);

var RemarkWrapper = exports.RemarkWrapper = _styledComponents2.default.span.withConfig({
    displayName: 'style__RemarkWrapper',
    componentId: 'uqtuoq-5'
})(['font-size:12px;line-height:24px;transition:all 0.3s;']);

var linkStatusMixin = function linkStatusMixin(_ref3) {
    var status = _ref3.status,
        DT = _ref3.theme.designTokens;

    if (status === 'before') {
        return (0, _styledComponents.css)(['color:', ';fill:', ';'], DT.T_COLOR_TEXT_PRIMARY_DEFAULT, DT.T_COLOR_TEXT_PRIMARY_DEFAULT);
    }
    return (0, _styledComponents.css)(['color:', ';fill:', ';'], DT.T_COLOR_TEXT_REMARK_DARK, DT.T_COLOR_TEXT_REMARK_DARK);
};

var stepStatusMixin = function stepStatusMixin(_ref4) {
    var status = _ref4.status,
        DT = _ref4.theme.designTokens;

    if (status === 'error') {
        return (0, _styledComponents.css)(['', ',', '{color:', ';}'], RemarkWrapper, TitleWrapper, DT.T_COLOR_TEXT_ERROR);
    }
    if (status === 'current' || status === 'before' || status === 'loading') {
        return (0, _styledComponents.css)(['', '{color:', ';}', '{color:', ';}'], RemarkWrapper, DT.T_COLOR_TEXT_DEFAULT_LIGHT, TitleWrapper, DT.T_COLOR_TEXT_DEFAULT_DARK);
    }
};

var StepWrapper = exports.StepWrapper = _styledComponents2.default.div.withConfig({
    displayName: 'style__StepWrapper',
    componentId: 'uqtuoq-6'
})(['', ';', ';'], _style.inlineBlockWithVerticalMixin, stepStatusMixin);

var LinkWrapper = exports.LinkWrapper = _styledComponents2.default.span.withConfig({
    displayName: 'style__LinkWrapper',
    componentId: 'uqtuoq-7'
})(['display:inline-block;text-align:center;line-height:32px;margin:0 32px;vertical-align:top;transition:all 0.3s;', ';'], linkStatusMixin);

var StepsWrapper = exports.StepsWrapper = _styledComponents2.default.div.withConfig({
    displayName: 'style__StepsWrapper',
    componentId: 'uqtuoq-8'
})(['']);

(0, _addDefaultThemeProps2.default)(StepsWrapper, IconWrapper, StepWrapper, LinkWrapper);
