// @ucloud-fe/react-components/lib/components/Message/MessageContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _uniqueId2 = require('lodash/uniqueId');

var _uniqueId3 = _interopRequireDefault(_uniqueId2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactTransitionGroup = require('react-transition-group');

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MessageContainer = (_temp2 = _class = function (_Component) {
    _inherits(MessageContainer, _Component);

    function MessageContainer() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, MessageContainer);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = MessageContainer.__proto__ || Object.getPrototypeOf(MessageContainer)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
            messages: [],
            top: _this.props.top
        }, _this.appendMessage = function (message) {
            var uid = (0, _uniqueId3.default)('uc_message_');
            _this.setState({
                messages: _this.state.messages.concat({
                    message: message,
                    uid: uid
                })
            });
            return uid;
        }, _this.removeMessage = function (uid) {
            var messages = _this.state.messages;

            var newMessages = messages.filter(function (info) {
                return info.uid !== uid;
            });
            var removed = newMessages.length !== messages.length;
            removed && _this.setState({
                messages: newMessages
            });
            return removed;
        }, _this.setTop = function (top) {
            _this.setState({
                top: top
            });
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(MessageContainer, [{
        key: 'render',
        value: function render() {
            var _state = this.state,
                messages = _state.messages,
                top = _state.top;
            // eslint-disable-next-line no-unused-vars

            var _props = this.props,
                _top = _props.top,
                rest = _objectWithoutProperties(_props, ['top']);

            return _react2.default.createElement(
                _style.MessageContentWrap,
                _extends({ style: { marginTop: top } }, rest),
                messages.map(function (info) {
                    return _react2.default.createElement(
                        _reactTransitionGroup.CSSTransition,
                        { key: info.uid, timeout: _style.animationDuration, classNames: _style.animationName },
                        info.message
                    );
                })
            );
        }
    }]);

    return MessageContainer;
}(_react.Component), _class.propTypes = {
    /** 顶部的间距 */
    top: _propTypes2.default.number
}, _temp2);
exports.default = MessageContainer;
