// @ucloud-fe/react-components/lib/__index.js

import * as BadgeAll from './components/Badge/';
const Badge = Object.assign(BadgeAll.default, BadgeAll);
export { Badge };

import * as BreadcrumbAll from './components/Breadcrumb/';
const Breadcrumb = Object.assign(BreadcrumbAll.default, BreadcrumbAll);
export { Breadcrumb };

import * as ButtonAll from './components/Button/';
const Button = Object.assign(ButtonAll.default, ButtonAll);
export { Button };

import * as CalendarAll from './components/Calendar/';
const Calendar = Object.assign(CalendarAll.default, CalendarAll);
export { Calendar };

import * as CardAll from './components/Card/';
const Card = Object.assign(CardAll.default, CardAll);
export { Card };

import * as CheckboxAll from './components/Checkbox/';
const Checkbox = Object.assign(CheckboxAll.default, CheckboxAll);
export { Checkbox };

import * as ColAll from './components/Col/';
const Col = Object.assign(ColAll.default, ColAll);
export { Col };

import * as CollapseAll from './components/Collapse/';
const Collapse = Object.assign(CollapseAll.default, CollapseAll);
export { Collapse };

import * as CombineAll from './components/Combine/';
const Combine = Object.assign(CombineAll.default, CombineAll);
export { Combine };

import * as CompactAll from './components/Compact/';
const Compact = Object.assign(CompactAll.default, CompactAll);
export { Compact };

import * as DatePickerAll from './components/DatePicker/';
const DatePicker = Object.assign(DatePickerAll.default, DatePickerAll);
export { DatePicker };

import * as DrawerAll from './components/Drawer/';
const Drawer = Object.assign(DrawerAll.default, DrawerAll);
export { Drawer };

import * as EditableListAll from './components/EditableList/';
const EditableList = Object.assign(EditableListAll.default, EditableListAll);
export { EditableList };

import * as EditableTableAll from './components/EditableTable/';
const EditableTable = Object.assign(EditableTableAll.default, EditableTableAll);
export { EditableTable };

import * as FormAll from './components/Form/';
const Form = Object.assign(FormAll.default, FormAll);
export { Form };

import * as GridAll from './components/Grid/';
const Grid = Object.assign(GridAll.default, GridAll);
export { Grid };

import * as IconAll from './components/Icon/';
const Icon = Object.assign(IconAll.default, IconAll);
export { Icon };

import * as InputAll from './components/Input/';
const Input = Object.assign(InputAll.default, InputAll);
export { Input };

import * as LoadingAll from './components/Loading/';
const Loading = Object.assign(LoadingAll.default, LoadingAll);
export { Loading };

import * as LocaleProviderAll from './components/LocaleProvider/';
const LocaleProvider = Object.assign(LocaleProviderAll.default, LocaleProviderAll);
export { LocaleProvider };

import * as MenuAll from './components/Menu/';
const Menu = Object.assign(MenuAll.default, MenuAll);
export { Menu };

import * as MessageAll from './components/Message/';
const Message = Object.assign(MessageAll.default, MessageAll);
export { Message };

import * as ModalAll from './components/Modal/';
const Modal = Object.assign(ModalAll.default, ModalAll);
export { Modal };

import * as NoticeAll from './components/Notice/';
const Notice = Object.assign(NoticeAll.default, NoticeAll);
export { Notice };

import * as NumberInputAll from './components/NumberInput/';
const NumberInput = Object.assign(NumberInputAll.default, NumberInputAll);
export { NumberInput };

import * as PaginationAll from './components/Pagination/';
const Pagination = Object.assign(PaginationAll.default, PaginationAll);
export { Pagination };

import * as PopConfirmAll from './components/PopConfirm/';
const PopConfirm = Object.assign(PopConfirmAll.default, PopConfirmAll);
export { PopConfirm };

import * as PopoverAll from './components/Popover/';
const Popover = Object.assign(PopoverAll.default, PopoverAll);
export { Popover };

import * as ProgressAll from './components/Progress/';
const Progress = Object.assign(ProgressAll.default, ProgressAll);
export { Progress };

import * as RadioAll from './components/Radio/';
const Radio = Object.assign(RadioAll.default, RadioAll);
export { Radio };

import * as RowAll from './components/Row/';
const Row = Object.assign(RowAll.default, RowAll);
export { Row };

import * as SelectAll from './components/Select/';
const Select = Object.assign(SelectAll.default, SelectAll);
export { Select };

import * as SliderAll from './components/Slider/';
const Slider = Object.assign(SliderAll.default, SliderAll);
export { Slider };

import * as SpinAll from './components/Spin/';
const Spin = Object.assign(SpinAll.default, SpinAll);
export { Spin };

import * as StepsAll from './components/Steps/';
const Steps = Object.assign(StepsAll.default, StepsAll);
export { Steps };

import * as SvgIconAll from './components/SvgIcon/';
const SvgIcon = Object.assign(SvgIconAll.default, SvgIconAll);
export { SvgIcon };

import * as SwitchAll from './components/Switch/';
const Switch = Object.assign(SwitchAll.default, SwitchAll);
export { Switch };

import * as TableAll from './components/Table/';
const Table = Object.assign(TableAll.default, TableAll);
export { Table };

import * as TabsAll from './components/Tabs/';
const Tabs = Object.assign(TabsAll.default, TabsAll);
export { Tabs };

import * as TagAll from './components/Tag/';
const Tag = Object.assign(TagAll.default, TagAll);
export { Tag };

import * as TextareaAll from './components/Textarea/';
const Textarea = Object.assign(TextareaAll.default, TextareaAll);
export { Textarea };

import * as ThemeProviderAll from './components/ThemeProvider/';
const ThemeProvider = Object.assign(ThemeProviderAll.default, ThemeProviderAll);
export { ThemeProvider };

import * as TooltipAll from './components/Tooltip/';
const Tooltip = Object.assign(TooltipAll.default, TooltipAll);
export { Tooltip };

import * as TransferAll from './components/Transfer/';
const Transfer = Object.assign(TransferAll.default, TransferAll);
export { Transfer };

import * as TransferMenuAll from './components/TransferMenu/';
const TransferMenu = Object.assign(TransferMenuAll.default, TransferMenuAll);
export { TransferMenu };

import * as TransferTableAll from './components/TransferTable/';
const TransferTable = Object.assign(TransferTableAll.default, TransferTableAll);
export { TransferTable };

import * as UploadAll from './components/Upload/';
const Upload = Object.assign(UploadAll.default, UploadAll);
export { Upload };

import * as ZFormAll from './components/ZForm/';
const ZForm = Object.assign(ZFormAll.default, ZFormAll);
export { ZForm };
