// @ucloud-fe/react-components/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZForm = exports.Upload = exports.TransferTable = exports.TransferMenu = exports.Transfer = exports.Tooltip = exports.ThemeProvider = exports.Textarea = exports.Tag = exports.Tabs = exports.Table = exports.Switch = exports.SvgIcon = exports.Steps = exports.Spin = exports.Slider = exports.Select = exports.Row = exports.Radio = exports.Progress = exports.Popover = exports.PopConfirm = exports.Pagination = exports.NumberInput = exports.Notice = exports.Modal = exports.Message = exports.Menu = exports.LocaleProvider = exports.Loading = exports.Input = exports.Icon = exports.Grid = exports.Form = exports.EditableTable = exports.EditableList = exports.Drawer = exports.DatePicker = exports.Compact = exports.Combine = exports.Collapse = exports.Col = exports.Checkbox = exports.Card = exports.Calendar = exports.Button = exports.Breadcrumb = exports.Badge = undefined;

var _Badge = require('./components/Badge/');

var BadgeAll = _interopRequireWildcard(_Badge);

var _Breadcrumb = require('./components/Breadcrumb/');

var BreadcrumbAll = _interopRequireWildcard(_Breadcrumb);

var _Button = require('./components/Button/');

var ButtonAll = _interopRequireWildcard(_Button);

var _Calendar = require('./components/Calendar/');

var CalendarAll = _interopRequireWildcard(_Calendar);

var _Card = require('./components/Card/');

var CardAll = _interopRequireWildcard(_Card);

var _Checkbox = require('./components/Checkbox/');

var CheckboxAll = _interopRequireWildcard(_Checkbox);

var _Col = require('./components/Col/');

var ColAll = _interopRequireWildcard(_Col);

var _Collapse = require('./components/Collapse/');

var CollapseAll = _interopRequireWildcard(_Collapse);

var _Combine = require('./components/Combine/');

var CombineAll = _interopRequireWildcard(_Combine);

var _Compact = require('./components/Compact/');

var CompactAll = _interopRequireWildcard(_Compact);

var _DatePicker = require('./components/DatePicker/');

var DatePickerAll = _interopRequireWildcard(_DatePicker);

var _Drawer = require('./components/Drawer/');

var DrawerAll = _interopRequireWildcard(_Drawer);

var _EditableList = require('./components/EditableList/');

var EditableListAll = _interopRequireWildcard(_EditableList);

var _EditableTable = require('./components/EditableTable/');

var EditableTableAll = _interopRequireWildcard(_EditableTable);

var _Form = require('./components/Form/');

var FormAll = _interopRequireWildcard(_Form);

var _Grid = require('./components/Grid/');

var GridAll = _interopRequireWildcard(_Grid);

var _Icon = require('./components/Icon/');

var IconAll = _interopRequireWildcard(_Icon);

var _Input = require('./components/Input/');

var InputAll = _interopRequireWildcard(_Input);

var _Loading = require('./components/Loading/');

var LoadingAll = _interopRequireWildcard(_Loading);

var _LocaleProvider = require('./components/LocaleProvider/');

var LocaleProviderAll = _interopRequireWildcard(_LocaleProvider);

var _Menu = require('./components/Menu/');

var MenuAll = _interopRequireWildcard(_Menu);

var _Message = require('./components/Message/');

var MessageAll = _interopRequireWildcard(_Message);

var _Modal = require('./components/Modal/');

var ModalAll = _interopRequireWildcard(_Modal);

var _Notice = require('./components/Notice/');

var NoticeAll = _interopRequireWildcard(_Notice);

var _NumberInput = require('./components/NumberInput/');

var NumberInputAll = _interopRequireWildcard(_NumberInput);

var _Pagination = require('./components/Pagination/');

var PaginationAll = _interopRequireWildcard(_Pagination);

var _PopConfirm = require('./components/PopConfirm/');

var PopConfirmAll = _interopRequireWildcard(_PopConfirm);

var _Popover = require('./components/Popover/');

var PopoverAll = _interopRequireWildcard(_Popover);

var _Progress = require('./components/Progress/');

var ProgressAll = _interopRequireWildcard(_Progress);

var _Radio = require('./components/Radio/');

var RadioAll = _interopRequireWildcard(_Radio);

var _Row = require('./components/Row/');

var RowAll = _interopRequireWildcard(_Row);

var _Select = require('./components/Select/');

var SelectAll = _interopRequireWildcard(_Select);

var _Slider = require('./components/Slider/');

var SliderAll = _interopRequireWildcard(_Slider);

var _Spin = require('./components/Spin/');

var SpinAll = _interopRequireWildcard(_Spin);

var _Steps = require('./components/Steps/');

var StepsAll = _interopRequireWildcard(_Steps);

var _SvgIcon = require('./components/SvgIcon/');

var SvgIconAll = _interopRequireWildcard(_SvgIcon);

var _Switch = require('./components/Switch/');

var SwitchAll = _interopRequireWildcard(_Switch);

var _Table = require('./components/Table/');

var TableAll = _interopRequireWildcard(_Table);

var _Tabs = require('./components/Tabs/');

var TabsAll = _interopRequireWildcard(_Tabs);

var _Tag = require('./components/Tag/');

var TagAll = _interopRequireWildcard(_Tag);

var _Textarea = require('./components/Textarea/');

var TextareaAll = _interopRequireWildcard(_Textarea);

var _ThemeProvider = require('./components/ThemeProvider/');

var ThemeProviderAll = _interopRequireWildcard(_ThemeProvider);

var _Tooltip = require('./components/Tooltip/');

var TooltipAll = _interopRequireWildcard(_Tooltip);

var _Transfer = require('./components/Transfer/');

var TransferAll = _interopRequireWildcard(_Transfer);

var _TransferMenu = require('./components/TransferMenu/');

var TransferMenuAll = _interopRequireWildcard(_TransferMenu);

var _TransferTable = require('./components/TransferTable/');

var TransferTableAll = _interopRequireWildcard(_TransferTable);

var _Upload = require('./components/Upload/');

var UploadAll = _interopRequireWildcard(_Upload);

var _ZForm = require('./components/ZForm/');

var ZFormAll = _interopRequireWildcard(_ZForm);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var Badge = Object.assign(BadgeAll.default, BadgeAll);
exports.Badge = Badge;

var Breadcrumb = Object.assign(BreadcrumbAll.default, BreadcrumbAll);
exports.Breadcrumb = Breadcrumb;

var Button = Object.assign(ButtonAll.default, ButtonAll);
exports.Button = Button;

var Calendar = Object.assign(CalendarAll.default, CalendarAll);
exports.Calendar = Calendar;

var Card = Object.assign(CardAll.default, CardAll);
exports.Card = Card;

var Checkbox = Object.assign(CheckboxAll.default, CheckboxAll);
exports.Checkbox = Checkbox;

var Col = Object.assign(ColAll.default, ColAll);
exports.Col = Col;

var Collapse = Object.assign(CollapseAll.default, CollapseAll);
exports.Collapse = Collapse;

var Combine = Object.assign(CombineAll.default, CombineAll);
exports.Combine = Combine;

var Compact = Object.assign(CompactAll.default, CompactAll);
exports.Compact = Compact;

var DatePicker = Object.assign(DatePickerAll.default, DatePickerAll);
exports.DatePicker = DatePicker;

var Drawer = Object.assign(DrawerAll.default, DrawerAll);
exports.Drawer = Drawer;

var EditableList = Object.assign(EditableListAll.default, EditableListAll);
exports.EditableList = EditableList;

var EditableTable = Object.assign(EditableTableAll.default, EditableTableAll);
exports.EditableTable = EditableTable;

var Form = Object.assign(FormAll.default, FormAll);
exports.Form = Form;

var Grid = Object.assign(GridAll.default, GridAll);
exports.Grid = Grid;

var Icon = Object.assign(IconAll.default, IconAll);
exports.Icon = Icon;

var Input = Object.assign(InputAll.default, InputAll);
exports.Input = Input;

var Loading = Object.assign(LoadingAll.default, LoadingAll);
exports.Loading = Loading;

var LocaleProvider = Object.assign(LocaleProviderAll.default, LocaleProviderAll);
exports.LocaleProvider = LocaleProvider;

var Menu = Object.assign(MenuAll.default, MenuAll);
exports.Menu = Menu;

var Message = Object.assign(MessageAll.default, MessageAll);
exports.Message = Message;

var Modal = Object.assign(ModalAll.default, ModalAll);
exports.Modal = Modal;

var Notice = Object.assign(NoticeAll.default, NoticeAll);
exports.Notice = Notice;

var NumberInput = Object.assign(NumberInputAll.default, NumberInputAll);
exports.NumberInput = NumberInput;

var Pagination = Object.assign(PaginationAll.default, PaginationAll);
exports.Pagination = Pagination;

var PopConfirm = Object.assign(PopConfirmAll.default, PopConfirmAll);
exports.PopConfirm = PopConfirm;

var Popover = Object.assign(PopoverAll.default, PopoverAll);
exports.Popover = Popover;

var Progress = Object.assign(ProgressAll.default, ProgressAll);
exports.Progress = Progress;

var Radio = Object.assign(RadioAll.default, RadioAll);
exports.Radio = Radio;

var Row = Object.assign(RowAll.default, RowAll);
exports.Row = Row;

var Select = Object.assign(SelectAll.default, SelectAll);
exports.Select = Select;

var Slider = Object.assign(SliderAll.default, SliderAll);
exports.Slider = Slider;

var Spin = Object.assign(SpinAll.default, SpinAll);
exports.Spin = Spin;

var Steps = Object.assign(StepsAll.default, StepsAll);
exports.Steps = Steps;

var SvgIcon = Object.assign(SvgIconAll.default, SvgIconAll);
exports.SvgIcon = SvgIcon;

var Switch = Object.assign(SwitchAll.default, SwitchAll);
exports.Switch = Switch;

var Table = Object.assign(TableAll.default, TableAll);
exports.Table = Table;

var Tabs = Object.assign(TabsAll.default, TabsAll);
exports.Tabs = Tabs;

var Tag = Object.assign(TagAll.default, TagAll);
exports.Tag = Tag;

var Textarea = Object.assign(TextareaAll.default, TextareaAll);
exports.Textarea = Textarea;

var ThemeProvider = Object.assign(ThemeProviderAll.default, ThemeProviderAll);
exports.ThemeProvider = ThemeProvider;

var Tooltip = Object.assign(TooltipAll.default, TooltipAll);
exports.Tooltip = Tooltip;

var Transfer = Object.assign(TransferAll.default, TransferAll);
exports.Transfer = Transfer;

var TransferMenu = Object.assign(TransferMenuAll.default, TransferMenuAll);
exports.TransferMenu = TransferMenu;

var TransferTable = Object.assign(TransferTableAll.default, TransferTableAll);
exports.TransferTable = TransferTable;

var Upload = Object.assign(UploadAll.default, UploadAll);
exports.Upload = Upload;

var ZForm = Object.assign(ZFormAll.default, ZFormAll);
exports.ZForm = ZForm;
