// @ucloud-fe/react-components/lib/components/Steps/Steps.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _style = require('./style');

var _Step = require('./Step');

var _Step2 = _interopRequireDefault(_Step);

var _Link = require('./Link');

var _Link2 = _interopRequireDefault(_Link);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Status = ['current', 'loading', 'error'];

var Steps = (_temp2 = _class = function (_Component) {
    _inherits(Steps, _Component);

    function Steps() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Steps);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Steps.__proto__ || Object.getPrototypeOf(Steps)).call.apply(_ref, [this].concat(args))), _this), _this.renderSteps = function (steps, current, status) {
            var pos = 'before';
            var l = steps.length;
            return steps.map(function (step, i) {
                var _step$key = step.key,
                    key = _step$key === undefined ? i : _step$key,
                    _step$step = step.step,
                    stepContent = _step$step === undefined ? i + 1 : _step$step,
                    rest = _objectWithoutProperties(step, ['key', 'step']);

                var isCurrent = current === key;

                var finalStatus = void 0;
                if (isCurrent) {
                    pos = 'after';
                    finalStatus = status;
                } else {
                    finalStatus = pos;
                }

                return [_react2.default.createElement(_Step2.default, _extends({}, rest, { key: key, status: finalStatus, step: stepContent })), i < l - 1 && _react2.default.createElement(_Link2.default, { key: 'link-' + key, status: pos })];
            });
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Steps, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                steps = _props.steps,
                current = _props.current,
                status = _props.status,
                rest = _objectWithoutProperties(_props, ['steps', 'current', 'status']);

            return _react2.default.createElement(
                _style.StepsWrapper,
                rest,
                this.renderSteps(steps, current, status)
            );
        }
    }]);

    return Steps;
}(_react.Component), _class.propTypes = {
    /** 步骤列表 */
    steps: _propTypes2.default.arrayOf(_propTypes2.default.shape({
        /** 步骤的 key，不传为数组索引 */
        key: _propTypes2.default.string,
        /** 步骤的显示文字，默认为索引 + 1 */
        step: _propTypes2.default.node,
        /** 步骤的标题内容 */
        title: _propTypes2.default.node,
        /** 步骤的备注 */
        remark: _propTypes2.default.node
    })).isRequired,
    /** 当前步骤的 key，不传或传 null 时为全部完成 */
    current: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
    /** 当前步骤的状态 */
    status: _propTypes2.default.oneOf(Status)
}, _class.defaultProps = {
    status: 'current'
}, _temp2);
exports.default = Steps;


Steps.Status = Status;
