// @ucloud-fe/react-components/lib/components/Upload/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.UploadNotice = exports.UploadIcon = exports.ListWrap = exports.TipWrap = exports.SelectorWrap = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _Notice = require('../../../components/Notice');

var _Notice2 = _interopRequireDefault(_Notice);

var _Icon = require('../../../components/Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var disabledMixin = (0, _styledComponents.css)(['pointer-events:none;']);

var SelectorWrap = exports.SelectorWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__SelectorWrap',
    componentId: 'efqsw2-0'
})(['display:inline-block;cursor:pointer;', ';'], function (_ref) {
    var disabled = _ref.disabled;
    return disabled && disabledMixin;
});
var TipWrap = exports.TipWrap = _styledComponents2.default.span.withConfig({
    displayName: 'style__TipWrap',
    componentId: 'efqsw2-1'
})(['margin-left:5px;']);

var ListWrap = exports.ListWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__ListWrap',
    componentId: 'efqsw2-2'
})(['max-height:200px;overflow:auto;border:1px solid ', ';margin-top:6px;'], function (_ref2) {
    var colorMap = _ref2.theme.colorMap;
    return colorMap.default.border;
});

var UploadIcon = exports.UploadIcon = (0, _styledComponents2.default)(_Icon2.default).withConfig({
    displayName: 'style__UploadIcon',
    componentId: 'efqsw2-3'
})(['vertical-align:baseline;margin-left:5px;cursor:pointer;']);

var UploadNotice = exports.UploadNotice = (0, _styledComponents2.default)(_Notice2.default).withConfig({
    displayName: 'style__UploadNotice',
    componentId: 'efqsw2-4'
})(['border:none;', ';'], function (_ref3) {
    var onPreview = _ref3.onPreview;
    return onPreview && (0, _styledComponents.css)(['cursor:pointer;']);
});
(0, _addDefaultThemeProps2.default)(ListWrap);
