// @ucloud-fe/react-components/lib/components/Card/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.CardWrap = exports.FooterWrap = exports.ContentWrap = exports.ActionWrap = exports.CommentWrap = exports.TitleWrap = exports.HeaderWrap = exports.prefixCls = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

var _style = require('../../../style');

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _prefixCls = _config2.default.prefixCls;
var prefixCls = exports.prefixCls = _prefixCls + '-card';

var sharedGutter = (0, _styledComponents.css)(['padding:0 24px;margin-top:16px;']);

var HeaderWrap = exports.HeaderWrap = _styledComponents2.default.div.attrs({
    className: prefixCls + '-header'
}).withConfig({
    displayName: 'style__HeaderWrap',
    componentId: 'yz23x1-0'
})(['', ';', ';'], _style.clearFixMixin, sharedGutter);

var TitleWrap = exports.TitleWrap = _styledComponents2.default.div.attrs({
    className: prefixCls + '-title'
}).withConfig({
    displayName: 'style__TitleWrap',
    componentId: 'yz23x1-1'
})(['line-height:28px;font-weight:bold;']);

var CommentWrap = exports.CommentWrap = _styledComponents2.default.div.attrs({
    className: prefixCls + '-comment'
}).withConfig({
    displayName: 'style__CommentWrap',
    componentId: 'yz23x1-2'
})(['line-height:20px;font-weight:normal;']);

var ActionWrap = exports.ActionWrap = _styledComponents2.default.div.attrs({
    className: prefixCls + '-action'
}).withConfig({
    displayName: 'style__ActionWrap',
    componentId: 'yz23x1-3'
})(['line-height:28px;', ';', ';'], _style.clearFixMixin, sharedGutter);

var ContentWrap = exports.ContentWrap = _styledComponents2.default.div.attrs({
    className: prefixCls + '-content'
}).withConfig({
    displayName: 'style__ContentWrap',
    componentId: 'yz23x1-4'
})(['', ';'], sharedGutter);

var FooterWrap = exports.FooterWrap = _styledComponents2.default.div.attrs({
    className: prefixCls + '-footer'
}).withConfig({
    displayName: 'style__FooterWrap',
    componentId: 'yz23x1-5'
})(['line-height:1;', ';padding-top:12px;', ';'], sharedGutter, _style.clearFixMixin);

/* stylelint-disable no-duplicate-selectors */
var themeMixin = function themeMixin(_ref) {
    var DT = _ref.theme.designTokens;
    return (0, _styledComponents.css)(['background:', ';font-size:', ';box-shadow:', ';border-radius:', ';', '{color:', ';font-size:', ';}', '{color:', ';font-size:', ';}', '{border-top:', ' solid ', ';padding:16px 24px;margin-top:24px;}'], DT.T_CARD_COLOR_BG_DEFAULT, DT.T_TYPO_FONT_SIZE_1, DT.T_SHADOW_BLOCK_DEFAULT_MD, DT.T_CORNER_LG, HeaderWrap, DT.T_COLOR_TEXT_DEFAULT_DARK, DT.T_TYPO_FONT_SIZE_3, CommentWrap, DT.T_COLOR_TEXT_DEFAULT_LIGHT, DT.T_TYPO_FONT_SIZE_1, FooterWrap, DT.T_LINE_WIDTH_BASE, DT.T_COLOR_LINE_DEFAULT_LIGHT);
};
/* stylelint-enable no-duplicate-selectors */

var CardWrap = exports.CardWrap = _styledComponents2.default.div.attrs({
    className: prefixCls
}).withConfig({
    displayName: 'style__CardWrap',
    componentId: 'yz23x1-6'
})(['box-sizing:border-box;text-align:left;overflow:auto;', ':last-child,', ':last-child,', ':last-child{margin-bottom:24px;}', ':first-child,', ':first-child,', ':first-child{margin-top:24px;}', ';'], HeaderWrap, ActionWrap, ContentWrap, HeaderWrap, ActionWrap, ContentWrap, themeMixin);

(0, _addDefaultThemeProps2.default)(CardWrap);
