// @ucloud-fe/react-components/lib/stores/SimpleStore.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var SimpleStore = function SimpleStore(initialState) {
    _classCallCheck(this, SimpleStore);

    return createStore(initialState);
};

exports.default = SimpleStore;


var createStore = function createStore(initialState) {
    var state = initialState;
    var listeners = [];

    function setState(partial) {
        state = Object.assign({}, state, partial);
    }

    function getState() {
        return state;
    }

    function dispatch() {
        for (var i = 0; i < listeners.length; i++) {
            var listener = listeners[i];
            listener();
        }
    }

    function subscribe(listener) {
        listeners.push(listener);

        return {
            unsubscribe: function unsubscribe() {
                return _unsubscribe(listener);
            }
        };
    }
    function _unsubscribe(listener) {
        var index = listeners.indexOf(listener);
        index >= 0 && listeners.splice(index, 1);
    }

    return {
        setState: setState,
        getState: getState,
        subscribe: subscribe,
        unsubscribe: _unsubscribe,
        dispatch: dispatch
    };
};
