// @ucloud-fe/react-components/lib/components/TransferTable/TransferTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Transfer = require('../../components/Transfer');

var _Transfer2 = _interopRequireDefault(_Transfer);

var _Table = require('../../components/Table');

var _Table2 = _interopRequireDefault(_Table);

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TransferTable = (_temp2 = _class = function (_PureComponent) {
    _inherits(TransferTable, _PureComponent);

    function TransferTable() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, TransferTable);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = TransferTable.__proto__ || Object.getPrototypeOf(TransferTable)).call.apply(_ref, [this].concat(args))), _this), _this.renderList = function (_ref2) {
            var dataSource = _ref2.dataSource,
                selectedKeys = _ref2.selectedKeys,
                onChange = _ref2.onChange,
                disabled = _ref2.disabled;
            var columns = _this.props.columns;

            return _react2.default.createElement(
                'div',
                { className: _style.tableWrapCls },
                _react2.default.createElement(_Table2.default, {
                    customStyle: { outerPadding: '0 12px 8px 12px' },
                    dataSource: dataSource,
                    rowSelection: {
                        selectedRowKeys: selectedKeys,
                        onChange: onChange,
                        disabled: disabled,
                        selectedTip: false
                    },
                    pagination: null,
                    columns: columns,
                    scroll: { y: 300 }
                })
            );
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(TransferTable, [{
        key: 'render',
        value: function render() {
            var rest = _objectWithoutProperties(this.props, []);

            return _react2.default.createElement(_Transfer2.default, _extends({}, rest, { renderList: this.renderList }));
        }
    }]);

    return TransferTable;
}(_react.PureComponent), _class.propTypes = Object.assign({}, _Transfer2.default.propTypes, {
    /** 表单的 columns */
    columns: _propTypes2.default.array
}), _temp2);
exports.default = TransferTable;
