// @ucloud-fe/react-components/lib/components/Calendar/Calendar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.calCalendarProps = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _localeConsumerDecorator = require('../../components/LocaleProvider/localeConsumerDecorator');

var _localeConsumerDecorator2 = _interopRequireDefault(_localeConsumerDecorator);

var _utils = require('./utils');

var _style = require('./style');

var _zh_CN = require('./locale/zh_CN');

var _zh_CN2 = _interopRequireDefault(_zh_CN);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Calendar = (_dec = (0, _localeConsumerDecorator2.default)({ defaultLocale: _zh_CN2.default, localeName: 'Calendar', publicFn: ['focus'] }), _dec(_class = (_temp2 = _class2 = function (_Component) {
    _inherits(Calendar, _Component);

    function Calendar() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Calendar);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Calendar.__proto__ || Object.getPrototypeOf(Calendar)).call.apply(_ref, [this].concat(args))), _this), _this.onSelect = function (value) {
            var _this$props = _this.props,
                rules = _this$props.rules,
                onSelect = _this$props.onSelect;

            value = (0, _utils.getValidDate)(value, rules);
            if (onSelect) {
                onSelect(value);
            }
        }, _this.focus = function () {
            _this.calendar && _this.calendar.focus();
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Calendar, [{
        key: 'render',
        value: function render() {
            var _this2 = this;

            /* eslint-disable-next-line no-unused-vars */
            var _props = this.props,
                _props$rules = _props.rules,
                rules = _props$rules === undefined ? {} : _props$rules,
                onSelect = _props.onSelect,
                rest = _objectWithoutProperties(_props, ['rules', 'onSelect']);

            return _react2.default.createElement(_style.CalendarWrap, _extends({
                _innerRef: function _innerRef(ref) {
                    return _this2.calendar = ref;
                },
                onSelect: this.onSelect
            }, calCalendarProps({ rules: rules }), rest));
        }
    }]);

    return Calendar;
}(_react.Component), _class2.propTypes = {
    /** 选中的时间，受控，Moment 类型 */
    value: _propTypes2.default.object,
    /** 默认选中的时间，非受控，Moment 类型  */
    defaultValue: _propTypes2.default.object,
    /** 实际选中时间修改的回调 */
    onSelect: _propTypes2.default.func,
    /** 变化时的回调，键盘操作快速切换等操作时触发 */
    onChange: _propTypes2.default.func,
    /** 自定义规则 */
    rules: _propTypes2.default.shape({
        /** 可选时间范围，格式为[start, end]，数据格式可以为moment实例、Date实例、时间戳 */
        range: _propTypes2.default.array,
        /**
         * 自定义禁用日期函数，返回true时该日期不可选
         * @param current - 日期
         * @param value - 日历当前选中值
         */
        custom: _propTypes2.default.func
    })
}, _class2.defaultProps = {}, _temp2)) || _class);


var calCalendarProps = function calCalendarProps(_ref2) {
    var rules = _ref2.rules;

    return {
        prefixCls: _style.prefixCls,
        showDateInput: false,
        showToday: false,
        disabledDate: function disabledDate(date, currentValue) {
            return (0, _utils.isDateDisabled)(date, currentValue, rules);
        }
    };
};

exports.default = Calendar;
exports.calCalendarProps = calCalendarProps;
