// @ucloud-fe/react-components/lib/components/Select/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.EmptyContentWrapper = exports.SelectWrap = exports.BlockMenu = exports.MenuWrap = exports.ExtraWrap = exports.OptionWrap = exports.Arrow = exports.Selector = exports.SelectSearchInput = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

var _Input = require('../../../components/Input');

var _Input2 = _interopRequireDefault(_Input);

var _Icon = require('../../../components/Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _Menu = require('../../../components/Menu');

var _Menu2 = _interopRequireDefault(_Menu);

var _Button = require('../../../components/Button');

var _Button2 = _interopRequireDefault(_Button);

var _style = require('../../../style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var SelectSearchInput = exports.SelectSearchInput = (0, _styledComponents2.default)(_Input2.default.Search).withConfig({
    displayName: 'style__SelectSearchInput',
    componentId: 'no8vyj-0'
})(['min-width:100px;display:block;margin:0 8px;margin-top:10px;']);

var Selector = exports.Selector = (0, _styledComponents2.default)(_Button2.default).withConfig({
    displayName: 'style__Selector',
    componentId: 'no8vyj-1'
})(['padding-right:28px;width:100%;text-align:left;']);

var Arrow = exports.Arrow = (0, _styledComponents2.default)(_Icon2.default).withConfig({
    displayName: 'style__Arrow',
    componentId: 'no8vyj-2'
})(['position:absolute;right:8px;top:50%;margin-top:-6px;']);
var OptionWrap = exports.OptionWrap = (0, _styledComponents2.default)(_Menu2.default.Item).withConfig({
    displayName: 'style__OptionWrap',
    componentId: 'no8vyj-3'
})(['', ';'], function (_ref) {
    var hidden = _ref.hidden;
    return hidden && (0, _styledComponents.css)(['display:none;']);
});
var ExtraWrap = exports.ExtraWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__ExtraWrap',
    componentId: 'no8vyj-4'
})(['margin:0 8px;']);
var MenuWrap = exports.MenuWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__MenuWrap',
    componentId: 'no8vyj-5'
})(function (_ref2) {
    var DT = _ref2.theme.designTokens;
    return (0, _styledComponents.css)(['box-shadow:', ';background:', ';border-radius:', ';display:inline-block;width:100%;& > ', ':last-child{margin-bottom:10px;}'], DT.T_SHADOW_BLOCK_DEFAULT_LG, DT.T_COLOR_BG_MENU, DT.T_CORNER_SM, /* sc-selector */ExtraWrap);
});

// eslint-disable-next-line react/prop-types,no-unused-vars
var CustomMenu = function CustomMenu(_ref3) {
    var customStyle = _ref3.customStyle,
        rest = _objectWithoutProperties(_ref3, ['customStyle']);

    return _react2.default.createElement(_Menu2.default, rest);
};

var BlockMenu = exports.BlockMenu = (0, _styledComponents2.default)(CustomMenu)(function (_ref4) {
    var customStyle = _ref4.customStyle;
    return (0, _styledComponents.css)(['display:block;border:none;box-shadow:none;max-height:', 'px;max-width:none;'], customStyle.optionListMaxHeight || 380);
});

/* stylelint-disable no-duplicate-selectors */
var propsMixin = function propsMixin(_ref5) {
    var DT = _ref5.theme.designTokens,
        disabled = _ref5.disabled;
    return (0, _styledComponents.css)(['font-size:', ';color:', ';', ';'], DT.T_TYPO_FONT_SIZE_1, DT.T_COLOR_TEXT_DEFAULT_DARK, disabled && (0, _styledComponents.css)(['color:', ';pointer-events:none;'], DT.T_COLOR_TEXT_DISABLED));
};
/* stylelint-enable no-duplicate-selectors */

var SelectWrap = exports.SelectWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__SelectWrap',
    componentId: 'no8vyj-6'
})(['box-sizing:border-box;position:relative;', ';', ';'], _style.inlineBlockWithVerticalMixin, propsMixin);

var EmptyContentWrapper = exports.EmptyContentWrapper = _styledComponents2.default.div.withConfig({
    displayName: 'style__EmptyContentWrapper',
    componentId: 'no8vyj-7'
})(function (_ref6) {
    var DT = _ref6.theme.designTokens;
    return (0, _styledComponents.css)(['text-align:center;color:', ';'], DT.T_COLOR_TEXT_REMARK_DARK);
});

(0, _addDefaultThemeProps2.default)(SelectWrap, MenuWrap, EmptyContentWrapper);
