// @ucloud-fe/react-components/lib/components/EditableTable/EditableTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Table = require('../../components/Table');

var _Table2 = _interopRequireDefault(_Table);

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var noop = function noop() {};

var EditableTable = (_temp2 = _class = function (_PureComponent) {
    _inherits(EditableTable, _PureComponent);

    function EditableTable() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, EditableTable);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = EditableTable.__proto__ || Object.getPrototypeOf(EditableTable)).call.apply(_ref, [this].concat(args))), _this), _this.onRemove = function (record) {
            var rowDeletion = _this.props.rowDeletion;
            var _rowDeletion$onDelete = rowDeletion.onDelete,
                onDelete = _rowDeletion$onDelete === undefined ? noop : _rowDeletion$onDelete;

            onDelete(record);
        }, _this.onAdd = function () {
            var addition = _this.props.addition;
            var _addition$onAdd = addition.onAdd,
                onAdd = _addition$onAdd === undefined ? noop : _addition$onAdd;

            onAdd();
        }, _this.renderFooter = function () {
            var _this$props = _this.props,
                footer = _this$props.footer,
                addition = _this$props.addition;
            var tip = addition.tip,
                disabled = addition.disabled;

            var _disabled = !addition || disabled;
            return [_react2.default.createElement(
                _style.AddBar,
                { key: 'add', onClick: !_disabled && _this.onAdd, disabled: _disabled },
                tip ? [_react2.default.createElement(_style.AddBtn, { type: 'plus', key: 'add_btn', size: '16px' }), _react2.default.createElement(
                    _style.AddTip,
                    { key: 'add_tip' },
                    tip
                )] : _react2.default.createElement(_style.AddBtn, { type: 'plus', size: '20px' })
            ), footer()];
        }, _this.getColumns = function () {
            var _this$props2 = _this.props,
                columns = _this$props2.columns,
                rowDeletion = _this$props2.rowDeletion;

            if (!rowDeletion) return columns;
            var getDisabledOfRow = rowDeletion.getDisabledOfRow,
                fixed = rowDeletion.fixed;

            var removeColumn = {
                title: null,
                key: '__editable_table_remove',
                width: 40,
                fixed: fixed,
                render: function render(v, record) {
                    return getDisabledOfRow && getDisabledOfRow(record) ? null : _react2.default.createElement(_style.RemoveBtn, { type: 'cross', onClick: function onClick() {
                            return _this.onRemove(record);
                        }, size: '16px' });
                }
            };
            return [].concat(_toConsumableArray(columns), [removeColumn]);
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(EditableTable, [{
        key: 'render',
        value: function render() {
            var rest = _objectWithoutProperties(this.props, []);

            return _react2.default.createElement(_Table2.default, _extends({}, rest, {
                pagination: null,
                footer: this.renderFooter,
                columns: this.getColumns(),
                emptyContent: null
            }));
        }
    }]);

    return EditableTable;
}(_react.PureComponent), _class.propTypes = {
    /**
     * @ignore
     */
    footer: _propTypes2.default.func,
    /**
     * @ignore
     */
    columns: _propTypes2.default.array.isRequired,
    /**
     * 是否可添加（false 时禁用添加栏）
     */
    addition: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.shape({
        /**
         * 添加回调
         */
        onAdd: _propTypes2.default.func,
        /**
         * 添加区域的自定义提示信息
         */
        tip: _propTypes2.default.node,
        /**
         * 是否禁用
         */
        disabled: _propTypes2.default.bool
    })]),
    /**
     * 是否可删除（false 隐藏删除按钮）
     */
    rowDeletion: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.shape({
        /**
         * 列删除回调
         * @param record 删除列数据
         */
        onDelete: _propTypes2.default.func,
        /**
         * 获取当前列的是否禁用删除
         * @param record 当前列数据
         */
        getDisabledOfRow: _propTypes2.default.func,
        /** 是否固定列 */
        fixed: _propTypes2.default.bool
    })])
}, _class.defaultProps = {
    footer: noop,
    addition: true,
    rowDeletion: true
}, _temp2);
exports.default = EditableTable;
