// @ucloud-fe/react-components/lib/components/Loading/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TipWrap = exports.ContentWrap = exports.LoadingIcon = exports.IndicatorWrap = exports.Mask = exports.LoadingWrap = exports.animationName = exports.animationDuration = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _Icon = require('../../../components/Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _animation = require('../../../style/animation');

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var animationDuration = exports.animationDuration = 500;
var animationName = exports.animationName = 'uc-fe-animation-fade';

var LoadingWrap = exports.LoadingWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__LoadingWrap',
    componentId: 'bpoj20-0'
})(['position:relative;.', '-enter,.', '-appear,.', '-leave,.', '-exit{animation-duration:', 'ms;animation-fill-mode:both;}.', '-enter,.', '-appear{animation-name:', ';}.', '-leave,.', '-exit{animation-name:', ';}'], animationName, animationName, animationName, animationName, animationDuration, animationName, animationName, _animation.fadeIn, animationName, animationName, _animation.fadeOut);

var Mask = exports.Mask = _styledComponents2.default.div.withConfig({
    displayName: 'style__Mask',
    componentId: 'bpoj20-1'
})(function (_ref) {
    var DT = _ref.theme.designTokens;
    return (0, _styledComponents.css)(['position:absolute;top:0;bottom:0;left:0;right:0;z-index:10;background:', ';'], DT.T_LOADING_COLOR_LAYER_DEFAULT);
});

var IndicatorWrap = exports.IndicatorWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__IndicatorWrap',
    componentId: 'bpoj20-2'
})(['display:table;width:100%;height:100%;& > div{display:table-cell;vertical-align:middle;text-align:center;}']);

var LoadingIcon = exports.LoadingIcon = (0, _styledComponents2.default)(_Icon2.default)(function (_ref2) {
    var DT = _ref2.theme.designTokens;
    return (0, _styledComponents.css)(['font-size:20px;color:', ';'], DT.T_COLOR_TEXT_PRIMARY_DEFAULT);
});

var ContentWrap = exports.ContentWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__ContentWrap',
    componentId: 'bpoj20-3'
})(['position:relative;']);

var TipWrap = exports.TipWrap = _styledComponents2.default.p.withConfig({
    displayName: 'style__TipWrap',
    componentId: 'bpoj20-4'
})(function (_ref3) {
    var _ref3$theme = _ref3.theme,
        fontSize = _ref3$theme.fontSize,
        DT = _ref3$theme.designTokens;
    return (0, _styledComponents.css)(['text-align:center;font-size:', ';color:', ';margin-top:5px;'], fontSize, DT.T_COLOR_TEXT_DEFAULT_DARK);
});

(0, _addDefaultThemeProps2.default)(TipWrap, LoadingIcon, Mask);
