// @ucloud-fe/react-components/lib/components/EditableList/EditableList.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _localeConsumerDecorator = require('../../components/LocaleProvider/localeConsumerDecorator');

var _localeConsumerDecorator2 = _interopRequireDefault(_localeConsumerDecorator);

var _Grid = require('../../components/Grid');

var _Button = require('../../components/Button');

var _Button2 = _interopRequireDefault(_Button);

var _SvgIcon = require('../../components/SvgIcon');

var _SvgIcon2 = _interopRequireDefault(_SvgIcon);

var _Combine = require('../../components/Combine');

var _Combine2 = _interopRequireDefault(_Combine);

var _Size = require('../../interfaces/Size');

var _Size2 = _interopRequireDefault(_Size);

var _zh_CN = require('./locale/zh_CN');

var _zh_CN2 = _interopRequireDefault(_zh_CN);

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var noop = function noop() {};

var colShape = {
    /** 栅格占位格数 */
    span: _propTypes2.default.number,
    /** 栅格向右偏移格数，偏移占位 */
    offset: _propTypes2.default.number,
    /** 栅格向左偏移格数，偏移不占位 */
    pull: _propTypes2.default.number,
    /** 栅格向右偏移格数，偏移不占位 */
    push: _propTypes2.default.number
};

var EditableList = (_dec = (0, _localeConsumerDecorator2.default)({ defaultLocale: _zh_CN2.default, localeName: 'EditableList' }), _dec(_class = (_temp2 = _class2 = function (_PureComponent) {
    _inherits(EditableList, _PureComponent);

    function EditableList() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, EditableList);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = EditableList.__proto__ || Object.getPrototypeOf(EditableList)).call.apply(_ref, [this].concat(args))), _this), _this.onAdd = function () {
            var addition = _this.props.addition;
            var _addition$onAdd = addition.onAdd,
                onAdd = _addition$onAdd === undefined ? noop : _addition$onAdd;

            onAdd();
        }, _this.onDelete = function (item) {
            var itemDeletion = _this.props.itemDeletion;
            var _itemDeletion$onDelet = itemDeletion.onDelete,
                onDelete = _itemDeletion$onDelet === undefined ? noop : _itemDeletion$onDelet;

            onDelete(item);
        }, _this.renderItem = function (item, latest) {
            var _this$props = _this.props,
                _this$props$renderIte = _this$props.renderItem,
                renderItem = _this$props$renderIte === undefined ? noop : _this$props$renderIte,
                itemDeletion = _this$props.itemDeletion,
                addition = _this$props.addition,
                grid = _this$props.grid,
                size = _this$props.size;
            var getDisabledOfItem = itemDeletion.getDisabledOfItem;

            var _item = renderItem(item);
            var _action = _react2.default.createElement(
                _Combine2.default,
                { sharedProps: { size: size } },
                !itemDeletion || getDisabledOfItem && getDisabledOfItem(item) ? null : _react2.default.createElement(
                    _Button2.default,
                    { onClick: function onClick() {
                            return _this.onDelete(item);
                        }, shape: 'square' },
                    _react2.default.createElement(_SvgIcon2.default, { type: 'minus' })
                ),
                latest && addition && !addition.disabled && _react2.default.createElement(
                    _Button2.default,
                    { onClick: _this.onAdd, shape: 'square' },
                    _react2.default.createElement(_SvgIcon2.default, { type: 'plus' })
                )
            );
            if (grid) {
                var _ref2 = grid || {},
                    _ref2$itemCol = _ref2.itemCol,
                    itemCol = _ref2$itemCol === undefined ? { span: 8 } : _ref2$itemCol,
                    _ref2$actionCol = _ref2.actionCol,
                    actionCol = _ref2$actionCol === undefined ? { span: 4 } : _ref2$actionCol;

                return _react2.default.createElement(
                    _Grid.Row,
                    { key: item.key, className: _style.liGridCls },
                    _react2.default.createElement(
                        _Grid.Col,
                        _extends({}, itemCol, { className: _style.itemCls }),
                        _item
                    ),
                    _react2.default.createElement(
                        _Grid.Col,
                        _extends({}, actionCol, { className: _style.actionCls }),
                        _action
                    )
                );
            }
            return _react2.default.createElement(
                'div',
                { key: item.key, className: _style.liCls },
                _react2.default.createElement(
                    'div',
                    { className: _style.itemCls },
                    _item
                ),
                _react2.default.createElement(
                    'div',
                    { className: _style.actionCls },
                    _action
                )
            );
        }, _this.renderList = function () {
            var _this$props2 = _this.props,
                _this$props2$dataSour = _this$props2.dataSource,
                dataSource = _this$props2$dataSour === undefined ? [] : _this$props2$dataSour,
                addition = _this$props2.addition,
                size = _this$props2.size,
                locale = _this$props2.locale;

            return !dataSource.length ? _react2.default.createElement(
                _Button2.default,
                { onClick: _this.onAdd, disabled: !addition || addition.disabled, size: size },
                locale.add
            ) : dataSource.map(function (item, i) {
                return _this.renderItem(item, i === dataSource.length - 1);
            });
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(EditableList, [{
        key: 'render',
        value: function render() {
            return _react2.default.createElement(
                _style.ListWrap,
                null,
                this.renderList()
            );
        }
    }]);

    return EditableList;
}(_react.PureComponent), _class2.propTypes = {
    /**
     * 数据列表
     */
    dataSource: _propTypes2.default.array,
    /**
     * 数据渲染
     * @param item 数据
     */
    renderItem: _propTypes2.default.func,
    /**
     * 使用 grid 布局
     */
    grid: _propTypes2.default.shape({
        /**
         * item 的 col 配置
         */
        itemCol: _propTypes2.default.shape(colShape),
        actionCol: _propTypes2.default.shape(colShape)
    }),
    /**
     * 自带控件尺寸
     */
    size: _propTypes2.default.oneOf(_Size2.default),
    /**
     * 是否可添加（false 时禁用添加栏）
     */
    addition: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.shape({
        /**
         * 添加回调
         */
        onAdd: _propTypes2.default.func,
        /**
         * 是否禁用
         */
        disabled: _propTypes2.default.bool
    })]),
    /**
     * 是否可删除（false 隐藏删除按钮）
     */
    itemDeletion: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.shape({
        /**
         * 列删除回调
         * @param record 删除列数据
         */
        onDelete: _propTypes2.default.func,
        /**
         * 获取当前列的是否禁用删除
         * @param record 当前列数据
         */
        getDisabledOfItem: _propTypes2.default.func
    })]),
    /** @ignore */
    locale: _propTypes2.default.object
}, _class2.defaultProps = {
    addition: true,
    itemDeletion: true,
    size: 'md'
}, _temp2)) || _class);
exports.default = EditableList;
