// @ucloud-fe/react-components/lib/components/Table/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.CancleSelect = exports.ActionButton = exports.ColumnConfigModalCheckbox = exports.ColumnConfigModalCheckboxGroup = exports.ColumnConfigModalNotice = exports.ColumnConfigButtonWrap = exports.ColumnConfigWrap = exports.PopupContainer = exports.TableWrap = exports.ExpandedRowContent = exports.placeholderHeaderCls = exports.placeholderCellCls = exports.selectIconHeaderCls = exports.selectIconCellCls = exports.SortIcon = exports.prefixCls = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _style = require('../../../style');

var _Button = require('../../../components/Button');

var _Button2 = _interopRequireDefault(_Button);

var _Notice = require('../../../components/Notice');

var _Notice2 = _interopRequireDefault(_Notice);

var _Checkbox = require('../../../components/Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _Icon = require('../../../components/Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _prefixCls = _config2.default.prefixCls;
var prefixCls = exports.prefixCls = _prefixCls + '-table';

var SortIcon = exports.SortIcon = (0, _styledComponents2.default)(_Icon2.default)(function (_ref) {
    var colorList = _ref.theme.colorList;
    return (0, _styledComponents.css)(['vertical-align:middle;color:', ';'], colorList.black);
});

var selectIconCellCls = exports.selectIconCellCls = prefixCls + '-row-select-icon-cell';
var selectIconHeaderCls = exports.selectIconHeaderCls = prefixCls + '-select-icon-th';

var placeholderCellCls = exports.placeholderCellCls = prefixCls + '-placeholder-cell';
var placeholderHeaderCls = exports.placeholderHeaderCls = prefixCls + '-placeholder-th';

var expandedRowContentCls = prefixCls + '-expanded-row-content';
var ExpandedRowContent = exports.ExpandedRowContent = _styledComponents2.default.div.attrs({
    className: expandedRowContentCls
})(function (_ref2) {
    var DT = _ref2.theme.designTokens;
    return (0, _styledComponents.css)(['color:', ';border-bottom:', ' solid ', ';background:', ';padding:12px;line-height:20px;'], DT.T_COLOR_TEXT_DEFAULT_LIGHT, DT.T_LINE_WIDTH_BASE, DT.T_COLOR_LINE_DEFAULT_LIGHT, DT.T_TABLE_ROW_COLOR_BG_DEFAULT);
});

var TableWrap = exports.TableWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__TableWrap',
    componentId: 'djz4q7-0'
})(function (_ref3) {
    var DT = _ref3.theme.designTokens,
        zebraCrossing = _ref3.zebraCrossing,
        _ref3$customStyle = _ref3.customStyle,
        customStyle = _ref3$customStyle === undefined ? {} : _ref3$customStyle;
    return (0, _styledComponents.css)(['', ';padding:', ';color:', ';border-radius:', ';background:', ';font-size:', ';.', '{&-pagination{float:right;margin-top:8px;}&-filter{border:none;padding:0;cursor:pointer;}table{border-collapse:separate;border-spacing:0;width:100%;min-width:100%;}&-filter-notice,&-empty-content-wrap,&-error-content-wrap{width:400px;margin:12px auto;text-align:center;}&-empty-content-wrap,&-error-content-wrap{margin:40px auto;}&-reset-link{color:', ';cursor:pointer;}&-fixed,&-body{background:', ';}&-row-expand-icon{color:', ';border:', ' solid ', ';', ';width:18px;height:18px;line-height:18px;text-align:center;cursor:pointer;&-cell{width:20px;}:hover{color:', ';border-color:', '}}&-expand-icon-th{width:20px;}&-row-expand-icon.', '-row-collapsed:before{content:\'+\';}&-row-expand-icon.', '-row-expanded:before{content:\'-\';}&-row-spaced{visibility:hidden;}&-row-expanded,&-row-collapsed{margin-right:8px;}&-row-expand-icon-cell .', '-row-expand-icon{margin:0;}', ' &-fixed-header .', '-header{min-width:100%;overflow:scroll;padding-bottom:20px;margin-bottom:-20px;}&-content{position:relative;width:100%;clear:both;}&-thead > tr > th{position:relative;vertical-align:middle;border-bottom:', ' solid ', ';color:', ';padding:12px;line-height:22px;text-align:left;font-weight:400;&.', '-expand-icon-th,&.', '{border-color:transparent;padding-left:0;}&.', '{padding:0px;font-size:0px;}}&-scroll-position-right,&-scroll-position-middle,&-scroll-position-left{.', '-th-fixed-left-latest,.', '-row-cell-fixed-left-latest,.', '-th-fixed-right-first,.', '-row-cell-fixed-right-first{::after{content:\'\';position:absolute;width:30px;height:100%;top:0;pointer-events:none;transition:box-shadow .3s;}}.', '-th-fixed-left-latest,.', '-row-cell-fixed-left-latest{::after{right:-30px;}}.', '-th-fixed-right-first,.', '-row-cell-fixed-right-first{::after{left:-30px;}}}&-scroll-position-right,&-scroll-position-middle{.', '-th-fixed-left-latest,.', '-row-cell-fixed-left-latest{::after{box-shadow:', ';}}}&-scroll-position-left,&-scroll-position-middle{.', '-th-fixed-right-first,.', '-row-cell-fixed-right-first{::after{box-shadow:', ';}}}&-row,&-row>td,&-thead,&-thead>tr,&-thead>tr>th{background:', ';}&-row > td{position:relative;vertical-align:middle;border-bottom:', ' solid ', ';padding:12px;line-height:22px;text-align:left;overflow-wrap:break-word;&.', '-row-expand-icon-cell,&.', '{border-color:transparent;padding-left:0;}&.', '{padding:0;font-size:0;}}&-row-level-1 > td,&-row-level-2 > td,&-row-level-3 > td,&-row-level-4 > td,&-row-level-5 > td{background:', ';&.', '-row-expand-icon-cell,&.', '-row-select-icon-cell{background:', ';}&.', '-row-expand-icon-cell .', '-row-spaced{visibility:visible;display:block;position:absolute;top:0;left:10px;width:1px;height:100%;background:', ';padding:0;border:none;border-bottom:', ' solid ', ';cursor:default;}}&-custom-title{margin-bottom:16px;}&-tip-wrap{border-bottom:', ' solid ', ';}table > tbody > .', '-row:hover > td{background:', ';&.', '-row-expand-icon-cell,&.', '{background:', ';}}&-row-indent.indent-level-0{display:none;}', ';}'], _style.clearFixMixin, customStyle.outerPadding || '15px', DT.T_COLOR_TEXT_DEFAULT_DARK, DT.T_CORNER_SM, DT.T_COLOR_BG_DEFAULT_NORMAL, DT.T_TYPO_FONT_SIZE_1, prefixCls, DT.T_COLOR_TEXT_PRIMARY_DEFAULT, DT.T_COLOR_BG_DEFAULT_NORMAL, DT.T_COLOR_TEXT_DEFAULT_LIGHT, DT.T_LINE_WIDTH_BASE, DT.T_COLOR_LINE_DEFAULT_LIGHT, _style.inlineBlockWithVerticalMixin, DT.T_COLOR_TEXT_PRIMARY_HOVER, DT.T_COLOR_LINE_PRIMARY_HOVER, prefixCls, prefixCls, prefixCls, function (_ref4) {
        var hideExpandIcon = _ref4.hideExpandIcon;
        return hideExpandIcon && (0, _styledComponents.css)(['&-row-expand-icon-cell,&-expand-icon-th,&-expand-icon-col,&-expanded-row > td:first-child{display:none;}']);
    }, prefixCls, DT.T_LINE_WIDTH_BASE, DT.T_COLOR_LINE_DEFAULT_DARK, DT.T_COLOR_TEXT_DEFAULT_LIGHT, prefixCls, selectIconHeaderCls, placeholderHeaderCls, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, prefixCls, DT.T_SHADOW_BLOCK_INSET_LEFT, prefixCls, prefixCls, DT.T_SHADOW_BLOCK_INSET_RIGHT, DT.T_COLOR_BG_DEFAULT_NORMAL, DT.T_LINE_WIDTH_BASE, DT.T_COLOR_LINE_DEFAULT_LIGHT, prefixCls, selectIconCellCls, placeholderCellCls, DT.T_TABLE_ROW_COLOR_BG_DEFAULT, prefixCls, prefixCls, DT.T_COLOR_BG_DEFAULT_NORMAL, prefixCls, prefixCls, DT.T_COLOR_LINE_DEFAULT_LIGHT, DT.T_LINE_WIDTH_BASE, DT.T_COLOR_LINE_DEFAULT_LIGHT, DT.T_LINE_WIDTH_BASE, DT.T_COLOR_LINE_DEFAULT_LIGHT, prefixCls, DT.T_TABLE_ROW_COLOR_BG_HOVER, prefixCls, selectIconCellCls, DT.T_COLOR_BG_DEFAULT_NORMAL, zebraCrossing && (0, _styledComponents.css)(['&-row:nth-child(odd) > td{background:', ';&.', '-row-expand-icon-cell,&.', '{background:', ';}}'], DT.T_TABLE_ROW_COLOR_BG_DEFAULT, prefixCls, selectIconCellCls, DT.T_COLOR_BG_DEFAULT_NORMAL));
});

var PopupContainer = exports.PopupContainer = _styledComponents2.default.div.withConfig({
    displayName: 'style__PopupContainer',
    componentId: 'djz4q7-1'
})(['position:relative;z-index:10;']);

var ColumnConfigWrap = exports.ColumnConfigWrap = _styledComponents2.default.span.withConfig({
    displayName: 'style__ColumnConfigWrap',
    componentId: 'djz4q7-2'
})(['', ';'], _style.inlineBlockWithVerticalMixin);

var ColumnConfigButtonWrap = exports.ColumnConfigButtonWrap = (0, _styledComponents2.default)(_Button2.default).withConfig({
    displayName: 'style__ColumnConfigButtonWrap',
    componentId: 'djz4q7-3'
})(['']);

var ColumnConfigModalNotice = exports.ColumnConfigModalNotice = (0, _styledComponents2.default)(_Notice2.default).withConfig({
    displayName: 'style__ColumnConfigModalNotice',
    componentId: 'djz4q7-4'
})(['border-width:0 0 1px 0;']);

var ColumnConfigModalCheckboxGroup = exports.ColumnConfigModalCheckboxGroup = (0, _styledComponents2.default)(_Checkbox2.default.Group).withConfig({
    displayName: 'style__ColumnConfigModalCheckboxGroup',
    componentId: 'djz4q7-5'
})(['padding:20px;']);

var ColumnConfigModalCheckbox = exports.ColumnConfigModalCheckbox = (0, _styledComponents2.default)(_Checkbox2.default).withConfig({
    displayName: 'style__ColumnConfigModalCheckbox',
    componentId: 'djz4q7-6'
})(['padding:12px 0;']);

var ActionButton = exports.ActionButton = (0, _styledComponents2.default)(_Button2.default).withConfig({
    displayName: 'style__ActionButton',
    componentId: 'djz4q7-7'
})(['margin-right:4px;']);

var CancleSelect = exports.CancleSelect = _styledComponents2.default.span.withConfig({
    displayName: 'style__CancleSelect',
    componentId: 'djz4q7-8'
})(function (_ref5) {
    var DT = _ref5.theme.designTokens;
    return (0, _styledComponents.css)(['cursor:pointer;color:', ';'], DT.T_COLOR_TEXT_PRIMARY_DEFAULT);
});

(0, _addDefaultThemeProps2.default)(TableWrap, ColumnConfigButtonWrap, SortIcon, ExpandedRowContent);
