// @ucloud-fe/react-components/lib/components/Progress/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.EndText = exports.TextWrap = exports.CurrentText = exports.Inner = exports.Outer = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _style = require('../../../style');

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Outer = exports.Outer = _styledComponents2.default.div.withConfig({
    displayName: 'style__Outer',
    componentId: 'sc-12bnnwr-0'
})(function (_ref) {
    var DT = _ref.theme.designTokens;
    return (0, _styledComponents.css)(['height:10px;background:', ';border-radius:5px;overflow:hidden;position:relative;'], DT.T_PROGRESS_COLOR_BG_DEFAULT);
});

var Inner = exports.Inner = _styledComponents2.default.div.withConfig({
    displayName: 'style__Inner',
    componentId: 'sc-12bnnwr-1'
})(function (_ref2) {
    var DT = _ref2.theme.designTokens,
        color = _ref2.color,
        percent = _ref2.percent;
    return (0, _styledComponents.css)(['width:', '%;height:100%;border:none;border-radius:5px;transition:width 0.5s;position:relative;background:', ';'], percent, color ? {
        success: DT.T_COLOR_BG_SUCCESS_DARK,
        warn: DT.T_COLOR_BG_WARNING_DARK,
        error: DT.T_COLOR_BG_ERROR_DARK,
        default: DT.T_COLOR_BG_PRIMARY_1
    }[color] || color : DT.T_COLOR_BG_PRIMARY_1);
});

var CurrentText = exports.CurrentText = _styledComponents2.default.span.withConfig({
    displayName: 'style__CurrentText',
    componentId: 'sc-12bnnwr-2'
})(function (_ref3) {
    var percent = _ref3.percent;
    return (0, _styledComponents.css)(['position:relative;left:', '%;width:50px;margin-left:-25px;display:inline-block;text-align:center;'], percent);
});

var TextWrap = exports.TextWrap = _styledComponents2.default.span.withConfig({
    displayName: 'style__TextWrap',
    componentId: 'sc-12bnnwr-3'
})(['position:relative;height:18px;line-height:18px;', ';'], _style.clearFixMixin);

var EndText = exports.EndText = _styledComponents2.default.span.withConfig({
    displayName: 'style__EndText',
    componentId: 'sc-12bnnwr-4'
})(['float:right;']);

(0, _addDefaultThemeProps2.default)(Inner, Outer);
