// @ucloud-fe/react-components/lib/components/PopConfirm/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.IconWrap = exports.FooterWrap = exports.ContentWrap = exports.PopupWrap = exports.prefixCls = undefined;

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

var _config = require('../../../config.js');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _prefixCls = _config2.default.prefixCls;
var prefixCls = exports.prefixCls = _prefixCls + '-popconfirm';

var PopupWrap = exports.PopupWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__PopupWrap',
    componentId: 'asjd35-0'
})(function () {
    return (0, _styledComponents.css)(['min-width:186px;max-width:300px;box-sizing:border-box;padding:12px 16px 12px 40px;']);
});
var ContentWrap = exports.ContentWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__ContentWrap',
    componentId: 'asjd35-1'
})(function (_ref) {
    var DT = _ref.theme.designTokens;
    return (0, _styledComponents.css)(['font-size:14px;font-weight:400;line-height:24px;max-height:72px;overflow:auto;word-break:normal;color:', ';'], DT.T_COLOR_TEXT_DEFAULT_DARK);
});
var FooterWrap = exports.FooterWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__FooterWrap',
    componentId: 'asjd35-2'
})(function () {
    return (0, _styledComponents.css)(['text-align:right;margin-top:12px;']);
});
var IconWrap = exports.IconWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__IconWrap',
    componentId: 'asjd35-3'
})(function (_ref2) {
    var DT = _ref2.theme.designTokens;
    return (0, _styledComponents.css)(['position:absolute;left:12px;top:14px;fill:', ';'], DT.T_COLOR_TEXT_WARNING);
});

(0, _addDefaultThemeProps2.default)(ContentWrap, IconWrap);
