// @ucloud-fe/react-components/lib/components/Grid/Col.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Col = function Col(_ref) {
    var rest = _objectWithoutProperties(_ref, []);

    return _react2.default.createElement(_style.ColWrap, rest);
};

Col.propTypes = {
    /** 栅格占位格数 */
    span: _propTypes2.default.number,
    /** 栅格向右偏移格数，偏移占位 */
    offset: _propTypes2.default.number,
    /** 栅格向左偏移格数，偏移不占位 */
    pull: _propTypes2.default.number,
    /** 栅格向右偏移格数，偏移不占位 */
    push: _propTypes2.default.number,
    /** flex布局下的order */
    order: _propTypes2.default.number,
    /** @ignore */
    gutter: _propTypes2.default.number,
    /** @ignore */
    className: _propTypes2.default.string
};

exports.default = Col;
