// @ucloud-fe/react-components/lib/components/Grid/Row.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _style = require('./style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Align = ['top', 'middle', 'bottom'];
var Justify = ['start', 'end', 'center', 'space-around', 'space-between'];
var Type = ['flex'];

var Row = function Row(_ref) {
    var _ref$gutter = _ref.gutter,
        gutter = _ref$gutter === undefined ? 16 : _ref$gutter,
        rest = _objectWithoutProperties(_ref, ['gutter']);

    return _react2.default.createElement(_style.RowWrap, _extends({ gutter: gutter }, rest));
};

Row.propTypes = {
    /** 是否为flex类型，flex类型时定位属性才能生效 */
    type: _propTypes2.default.oneOf(Type),
    /** 垂直定位 */
    align: _propTypes2.default.oneOf(Align),
    /** 水平定位 */
    justify: _propTypes2.default.oneOf(Justify),
    /** 栅格间距 */
    gutter: _propTypes2.default.number,
    /** @ignore */
    className: _propTypes2.default.string,
    /** @ignore */
    style: _propTypes2.default.object,
    /** @ignore */
    children: _propTypes2.default.node
};

Object.assign(Row, {
    Align: Align,
    Justify: Justify,
    Type: Type
});

exports.default = Row;
