// @ucloud-fe/react-components/lib/components/Menu/style/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.PopupMenuWrap = exports.MenuWrap = exports.SubMenuIcon = exports.checkboxCls = exports.popupContentCls = exports.popupWrapCls = exports.popupTitleCls = exports.collapseWrapCls = exports.collapseTitleCls = exports.selectallWrapCls = exports.selectedCls = exports.lastCls = exports.firstCls = exports.hiddenCls = exports.disabledCls = exports.itemCls = exports.multipleCls = exports.singleCls = exports.blockCls = exports.prefixCls = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _Collapse = require('../../../components/Collapse');

var _Collapse2 = _interopRequireDefault(_Collapse);

var _Icon = require('../../../components/Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _config = require('../../../config');

var _config2 = _interopRequireDefault(_config);

var _addDefaultThemeProps = require('../../../components/ThemeProvider/addDefaultThemeProps');

var _addDefaultThemeProps2 = _interopRequireDefault(_addDefaultThemeProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _prefixCls = _config2.default.prefixCls;
var prefixCls = exports.prefixCls = _prefixCls + '-menu';
var blockCls = exports.blockCls = _prefixCls + '-block';
var singleCls = exports.singleCls = prefixCls + '-single';
var multipleCls = exports.multipleCls = prefixCls + '-multiple';
var itemCls = exports.itemCls = prefixCls + '-item';
var disabledCls = exports.disabledCls = itemCls + '-disabled';
var hiddenCls = exports.hiddenCls = itemCls + '-hidden';
var firstCls = exports.firstCls = prefixCls + '-first';
var lastCls = exports.lastCls = prefixCls + '-last';
var selectedCls = exports.selectedCls = prefixCls + '-selected';
var selectallWrapCls = exports.selectallWrapCls = prefixCls + '-selectall-wrap';
var collapseTitleCls = exports.collapseTitleCls = prefixCls + '-collapse-title';
var collapseWrapCls = exports.collapseWrapCls = prefixCls + '-collapse-wrap';
var popupTitleCls = exports.popupTitleCls = prefixCls + '-popup-title';
var popupWrapCls = exports.popupWrapCls = prefixCls + '-popup-wrap';
var popupContentCls = exports.popupContentCls = prefixCls + '-popup-content';
var checkboxCls = exports.checkboxCls = prefixCls + '-checkbox';

var SubMenuIcon = exports.SubMenuIcon = (0, _styledComponents2.default)(_Icon2.default).withConfig({
    displayName: 'style__SubMenuIcon',
    componentId: 'ulx11w-0'
})(['position:absolute;right:8px;top:50%;margin-top:-0.5em;font-size:14px;']);

var menuStyle = function menuStyle(_ref) {
    var DT = _ref.theme.designTokens;

    return (0, _styledComponents.css)(['display:inline-block;box-sizing:border-box;overflow:auto;line-height:32px;font-size:12px;max-width:360px;min-width:64px;padding:4px 0px;border-radius:2px;text-align:left;border-style:none !important;border:none;color:', ';box-shadow:', ';background:', ';.', ',.', ',.', ',.', '{cursor:pointer;white-space:nowrap;text-decoration:none;text-overflow:ellipsis;overflow:hidden;:hover{background:', ';}}.', ',.', ',.', ',.', '{padding:0 8px;margin:0 8px;}.', '.', ',.', '.', '{color:', ';cursor:not-allowed;:hover{background:none;}}.', '{display:none;}.', ',.', '{position:relative;padding:0px 40px 0px 8px;&.', '{color:', ';}}&.', ' .', '.', '{color:', ';}.', '{::after,::before{content:\'\';display:block;margin:4px 8px;height:1px;background:', ';}}.', '.', '{::before{display:none;}}.', '+.', '{::before{display:none;}}.', '.', '{::after{display:none;}}.', '{width:100%;}'], DT.T_COLOR_TEXT_DEFAULT_DARK, DT.T_SHADOW_BLOCK_DEFAULT_LG, DT.T_COLOR_BG_MENU, itemCls, selectallWrapCls, popupTitleCls, collapseTitleCls, DT.T_COLOR_BG_DEFAULT_HOVER, itemCls, selectallWrapCls, popupTitleCls, collapseTitleCls, itemCls, disabledCls, selectallWrapCls, disabledCls, DT.T_COLOR_TEXT_DISABLED, hiddenCls, popupTitleCls, collapseTitleCls, selectedCls, DT.T_COLOR_TEXT_PRIMARY_DEFAULT, singleCls, itemCls, selectedCls, DT.T_COLOR_TEXT_PRIMARY_DEFAULT, collapseWrapCls, DT.T_COLOR_LINE_DEFAULT_LIGHT, firstCls, collapseWrapCls, collapseWrapCls, collapseWrapCls, lastCls, collapseWrapCls, checkboxCls);
};

// eslint-disable-next-line react/prop-types,no-unused-vars
var CleanCollapse = function CleanCollapse(_ref2) {
    var customStyle = _ref2.customStyle,
        rest = _objectWithoutProperties(_ref2, ['customStyle']);

    return _react2.default.createElement(_Collapse2.default, rest);
};

var MenuWrap = exports.MenuWrap = (0, _styledComponents2.default)(CleanCollapse)(function (props) {
    var _props$customStyle = props.customStyle,
        customStyle = _props$customStyle === undefined ? {} : _props$customStyle;

    return (0, _styledComponents.css)(['', ';max-height:', ';&.', '{min-width:0;max-width:none;width:100%;height:100%;box-shadow:none;}'], menuStyle(props), customStyle.maxHeight, blockCls);
});

var PopupMenuWrap = exports.PopupMenuWrap = _styledComponents2.default.div.withConfig({
    displayName: 'style__PopupMenuWrap',
    componentId: 'ulx11w-1'
})(function (props) {
    return (0, _styledComponents.css)(['display:inline-block;padding:0px 8px;.', '{', ';max-height:380px;}'], popupContentCls, menuStyle(props));
});

(0, _addDefaultThemeProps2.default)(MenuWrap, PopupMenuWrap);
